{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "89b274b8-2cf4-48b4-bb42-9f65a079fbc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from narrative_llm_agent.crews.job_crew import JobCrew\n",
    "from narrative_llm_agent.crews.startup_crew import StartupCrew\n",
    "from langchain_openai import ChatOpenAI\n",
    "import os\n",
    "from crewai import Crew, Agent, Task\n",
    "from narrative_llm_agent.agents.workflow import WorkflowRunner\n",
    "from narrative_llm_agent.agents.job import AppOutputInfo\n",
    "from langchain.tools import tool\n",
    "import json\n",
    "from pydantic import BaseModel, Field, PrivateAttr\n",
    "from typing import Type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3189fabe-8064-4035-b71e-c188f280d7cd",
   "metadata": {},
   "source": [
    "# Combine the intro (get object and metadata) and app running workflows.\n",
    "Should do the following, in order.\n",
    "1. X User indepedently makes a new narrative and adds a reads object or two to it.\n",
    "2. X User notes narrative id. (possible we can have the agent fetch the most recent few, start with narrative id)\n",
    "3. X Run the startup crew\n",
    "4. X Should return object with narrative id, reads, metadata\n",
    "5. (later) Run the Analyst, set up apps to run (for now, use curated set from Prachi)\n",
    "6. Run Workflow Manager over them all."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d20254ef-fb01-40da-9759-eb54e13e81b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "narrative_id = 211157"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8138f84f-e30a-4be9-89fc-d2c7ba35e6ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm = ChatOpenAI(\n",
    "    model=\"openai/gpt-4o\",\n",
    "    api_key=os.environ.get(\"CBORG_API_KEY\"),\n",
    "    base_url=\"https://api.cborg.lbl.gov\",  # For LBL-Net, use \"https://api-local.cborg.lbl.gov\"\n",
    "    temperature=0,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f631ee5-699a-468c-8114-4fd59151b90e",
   "metadata": {},
   "outputs": [],
   "source": [
    "startup_crew = StartupCrew(llm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bb6798f0-b551-4e9f-b5cb-b4f80ec3031a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Ask the user which narrative id they are using. This will be a number. Return only that number for the next task.\n",
      "        \u001b[00m\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mAsk the user which narrative id they are using. This will be a number. Return only that number for the next task.\u001b[00m\n",
      "Could you please provide the narrative id you are using? It should be a number.\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " 211157\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mThought: I need to ask the user for the narrative id they are using, which is a number, to proceed with the next task.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mconversation-tool\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"argument\\\": \\\"Could you please provide the narrative id you are using? It should be a number.\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "211157\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "211157\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mDelegate work to coworker\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"task\\\": \\\"Ask the user which narrative id they are using. This will be a number. Return only that number for the next task.\\\", \\\"context\\\": \\\"We need to obtain the numeric narrative id from the user to proceed with the next task. This id is crucial for the project's workflow.\\\", \\\"coworker\\\": \\\"Human Interaction Manager\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "211157\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "211157\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Fetch the list of all reads objects available in the user's narrative. The narrative id is a number available\n",
      "            in teh task context. Do not ask the user for the narrative id. Do not ask the user for the list of reads\n",
      "            directly, use a tool. Return the UPA and name for each of the reads objects.\n",
      "        \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mI need to fetch the list of all objects available in the user's narrative using the provided narrative ID. Then, I will filter out the reads objects and retrieve their names and UPAs.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mlist-objects\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "[{\"ws_id\": 211157, \"obj_id\": 1, \"name\": \"Narrative.1742916273434\", \"ws_name\": \"wjriehl:narrative_1742916273434\", \"metadata\": null, \"type\": \"KBaseNarrative.Narrative-4.0\", \"saved\": \"2025-03-25T17:09:57+0000\", \"version\": 4, \"saved_by\": \"wjriehl\", \"size_bytes\": 2079, \"upa\": \"211157/1/4\"}, {\"ws_id\": 211157, \"obj_id\": 2, \"name\": \"rhodo.art.q20.PE.reads\", \"ws_name\": \"wjriehl:narrative_1742916273434\", \"metadata\": null, \"type\": \"KBaseFile.PairedEndLibrary-2.1\", \"saved\": \"2025-03-25T15:25:43+0000\", \"version\": 1, \"saved_by\": \"wjriehl\", \"size_bytes\": 765, \"upa\": \"211157/2/1\"}]\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"reads\": [{\n",
      "    \"narrative_id\": 211157,\n",
      "    \"reads_object_name\": \"rhodo.art.q20.PE.reads\",\n",
      "    \"reads_object_upa\": \"211157/2/1\"\n",
      "  }]\n",
      "}\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            From the list of available reads, ask the user with set of reads they want to assemble and annotate.\n",
      "            Return the narrative id, the UPA of the user's chosen reads object, and the name of the reads object.\n",
      "        \u001b[00m\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mAsk the user which set of reads they want to assemble and annotate from the available list.\u001b[00m\n",
      "Which set of reads would you like to assemble and annotate? The available option is 'rhodo.art.q20.PE.reads'.\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " Yes, use the only one available.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mI need to ask the user which set of reads they want to assemble and annotate from the available list.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mconversation-tool\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"argument\\\": \\\"Which set of reads would you like to assemble and annotate? The available option is 'rhodo.art.q20.PE.reads'.\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "Yes, use the only one available.\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "The user has chosen to assemble and annotate the reads set 'rhodo.art.q20.PE.reads'.\u001b[00m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mDelegate work to coworker\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"task\\\": \\\"Ask the user which set of reads they want to assemble and annotate from the available list.\\\", \\\"context\\\": \\\"The available reads are: [{'narrative_id': 211157, 'reads_object_name': 'rhodo.art.q20.PE.reads', 'reads_object_upa': '211157/2/1'}]. We need to know the user's choice to proceed with the assembly and annotation.\\\", \\\"coworker\\\": \\\"Human Interaction Manager\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "The user has chosen to assemble and annotate the reads set 'rhodo.art.q20.PE.reads'.\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\"narrative_id\": 211157, \"reads_object_name\": \"rhodo.art.q20.PE.reads\", \"reads_object_upa\": \"211157/2/1\"}\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Assemble metadata necessary to assemble and annotate the genomic reads in the UPA given by context.\n",
      "            First, retrieve the object metadata for that UPA.\n",
      "            If there is not enough information here to choose appropriate applications for assembly and annotation, then ask the user all necessary questions until enough information is together.\n",
      "            Note that the user may not know certain information, and this is a valid answer.\n",
      "            Assemble the final results into a JSON string and store this conversation in the from the context.\n",
      "            Return the final results.\n",
      "        \u001b[00m\n",
      "looking up obj info for 211157/2/1\n",
      "got object info\n",
      "{'ws_id': 211157, 'obj_id': 2, 'name': 'rhodo.art.q20.PE.reads', 'ws_name': 'wjriehl:narrative_1742916273434', 'metadata': None, 'type': 'KBaseFile.PairedEndLibrary-2.1', 'saved': '2025-03-25T15:25:43+0000', 'version': 1, 'saved_by': 'wjriehl', 'size_bytes': 765, 'upa': '211157/2/1'}\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mget-object-metadata\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"obj_upa\\\": \\\"211157/2/1\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "null\u001b[00m\n",
      "Hello! I need to gather some information about your genomic reads to proceed with the assembly and annotation. Could you please provide details on the organism, read length, insert size, library type, whether the reads are paired-end, sequencing platform, sequencing depth, genome size, ploidy, and coverage? Additionally, if you have any preferred assembler, annotator, or QC tools, please let me know. If you're unsure about any of these, that's perfectly fine too!\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " The reads are paired-end, and they come from a Rhodobacter species.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mconversation-tool\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"argument\\\": \\\"Hello! I need to gather some information about your genomic reads to proceed with the assembly and annotation. Could you please provide details on the organism, read length, insert size, library type, whether the reads are paired-end, sequencing platform, sequencing depth, genome size, ploidy, and coverage? Additionally, if you have any preferred assembler, annotator, or QC tools, please let me know. If you're unsure about any of these, that's perfectly fine too!\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "The reads are paired-end, and they come from a Rhodobacter species.\u001b[00m\n",
      "Great, thank you! Could you also provide the read length, insert size, library type, sequencing platform, sequencing depth, genome size, ploidy, and coverage? Additionally, if you have any preferred assembler, annotator, or QC tools, please let me know. If you're unsure about any of these, that's perfectly fine too!\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " I have no preference.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mconversation-tool\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"argument\\\": \\\"Great, thank you! Could you also provide the read length, insert size, library type, sequencing platform, sequencing depth, genome size, ploidy, and coverage? Additionally, if you have any preferred assembler, annotator, or QC tools, please let me know. If you're unsure about any of these, that's perfectly fine too!\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "I have no preference.\n",
      "\n",
      "\n",
      "You ONLY have access to the following tools, and should NEVER make up tools that are not listed here:\n",
      "\n",
      "Tool Name: conversation-tool\n",
      "Tool Arguments: {'argument': {'description': None, 'type': 'str'}}\n",
      "Tool Description: Converse with the user until all questions are fully answered.\n",
      "Tool Name: get-object-metadata\n",
      "Tool Arguments: {'obj_upa': {'description': None, 'type': 'str'}}\n",
      "Tool Description: Return the metadata for a KBase Workspace object with the given UPA.\n",
      "Tool Name: store-conversation\n",
      "Tool Arguments: {'narrative_id': {'description': 'id of the Narrative to store the conversation in', 'type': 'int'}, 'json_conversation': {'description': 'JSON format of the conversation to store', 'type': 'str'}}\n",
      "Tool Description: Store Conversation Tool. This securely stores the results of a conversation in a KBase Narrative.\n",
      "\n",
      "IMPORTANT: Use the following format in your response:\n",
      "\n",
      "```\n",
      "Thought: you should always think about what to do\n",
      "Action: the action to take, only one name of [conversation-tool, get-object-metadata, store-conversation], just the name, exactly as it's written.\n",
      "Action Input: the input to the action, just a simple JSON object, enclosed in curly braces, using \" to wrap keys and values.\n",
      "Observation: the result of the action\n",
      "```\n",
      "\n",
      "Once all necessary information is gathered, return the following format:\n",
      "\n",
      "```\n",
      "Thought: I now know the final answer\n",
      "Final Answer: the final answer to the original input question\n",
      "```\u001b[00m\n",
      "Thank you for your patience. Could you please provide the read length, insert size, library type, sequencing platform, sequencing depth, genome size, ploidy, and coverage? If you're unsure about any of these, that's perfectly fine too!\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " I don't know any of those.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mconversation-tool\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"argument\\\": \\\"Thank you for your patience. Could you please provide the read length, insert size, library type, sequencing platform, sequencing depth, genome size, ploidy, and coverage? If you're unsure about any of these, that's perfectly fine too!\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "I don't know any of those.\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mHuman Interaction Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"organism\": \"Rhodobacter species\",\n",
      "  \"read_length\": null,\n",
      "  \"insert_size\": null,\n",
      "  \"library_type\": null,\n",
      "  \"paired_end\": true,\n",
      "  \"platform\": null,\n",
      "  \"sequencing_depth\": null,\n",
      "  \"genome_size\": null,\n",
      "  \"ploidy\": null,\n",
      "  \"coverage\": null,\n",
      "  \"additional_notes\": null,\n",
      "  \"preferred_assembler\": null,\n",
      "  \"preferred_annotator\": null,\n",
      "  \"preferred_qc_tools\": null\n",
      "}\n",
      "```\u001b[00m\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "result = startup_crew.run_startup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4c8719cf-2167-47a5-b450-c71637957e83",
   "metadata": {},
   "outputs": [],
   "source": [
    "wf_runner = WorkflowRunner(llm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e6844d78-b063-4ad4-a747-751ed822158a",
   "metadata": {},
   "outputs": [],
   "source": [
    "steps_to_run = [\n",
    "    {\n",
    "        \"Step\": 1,\n",
    "        \"Name\": \"Quality control\",\n",
    "        \"Description\": \"Check reads for quality using FastQC\",\n",
    "        \"expect_new_object\": False,\n",
    "        \"app_id\": \"kb_fastqc/runFastQC\"\n",
    "    }, \n",
    "    {\n",
    "        \"Step\": 2,\n",
    "        \"Name\": \"Reads trimming\",\n",
    "        \"Description\": \"Trim low-quality bases and remove adapter sequences from the paired-end reads using Trimmomatic\",\n",
    "        \"expect_new_object\": True,\n",
    "        \"app_id\": \"kb_trimmomatic/run_trimmomatic\"\n",
    "    },\n",
    "    {\n",
    "        \"Step\": 3,\n",
    "        \"Name\": \"Genome Assembly\",\n",
    "        \"Description\": \"Perform de novo assembly of the trimmed reads into contigs and scaffolds using SPAdes.\",\n",
    "        \"expect_new_object\": True,\n",
    "        \"app_id\": \"kb_SPAdes/run_SPAdes\"\n",
    "    },\n",
    "    {\n",
    "        \"Step\": 4,\n",
    "        \"Name\": \"Assembly Quality Assessment\",\n",
    "        \"Description\": \"Obtain assembly statistics such as N50, total length, and number of contigs using QUAST.\", \n",
    "        \"expect_new_object\": False,\n",
    "        \"app_id\": \"kb_quast/run_QUAST_app\"\n",
    "    },\n",
    "    {\n",
    "        \"Step\": 5, \n",
    "        \"Name\": \"Genome Annotation\", \n",
    "        \"Description\": \"Annotate the assembled genome using Prokka to provide insights into gene content and functional elements.\", \n",
    "        \"expect_new_object\": True,\n",
    "        \"app_id\": \"ProkkaAnnotation/annotate_contigs\"\n",
    "    },\n",
    "    {\n",
    "        \"Step\": 6, \n",
    "        \"Name\": \"Genome Completeness Estimation\", \n",
    "        \"Description\": \"Estimate genome completeness and contamination levels using CheckM.\", \n",
    "        \"expect_new_object\": False,\n",
    "        \"app_id\": \"kb_Msuite/run_checkM_lineage_wf\"\n",
    "    },\n",
    "    {\n",
    "        \"Step\": 7,\n",
    "        \"Name\": \"Build Genome Set\",\n",
    "        \"Description\": \"Build a Genome Set to run through GTDB-tk\",\n",
    "        \"expect_new_object\": True,\n",
    "        \"app_id\": \"kb_SetUtilities/KButil_Build_GenomeSet\",\n",
    "    },\n",
    "    {\n",
    "        \"Step\": 8, \n",
    "        \"Name\": \"Taxonomic Classification\", \n",
    "        \"Description\": \"Perform taxonomic classification to confirm organism identity using GTDB-Tk.\", \n",
    "        \"app_id\": \"kb_gtdbtk/run_kb_gtdbtk_classify_wf\",\n",
    "        \"expect_new_object\": True\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9eaa847d-2f4a-4782-8d12-a7920684f4f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_apps_task = Task(\n",
    "    description=f\"\"\"\n",
    "    This task involves running multiple apps where the output of one (if any) is fed into the next as input. \n",
    "    Here are the tasks in JSON format: {json.dumps(steps_to_run)}.\n",
    "    If any task has \"expect_new_object\" set to True, then that should receive a new data object in its output as a \"created_object\". That object should be used as input for the next task.\n",
    "    If a task as \"expect_new_object\" set to False, then that should not receive a new object to use in the next task. In that case, use the same input object from the previous step for the next one.\n",
    "    These steps must be run sequentially. \n",
    "    These must be run in the narrative with id {result.narrative_id} and start with using the paired-end reads object {result.reads_object_upa} with name {result.reads_object_name}.\n",
    "    If any step ends with an error, immediately stop the task and end with an error.\n",
    "    In the end, return a brief summary of steps taken and resulting output objects.\n",
    "    \"\"\",\n",
    "    expected_output=\"A summary of task completion, the number of apps run, and the upa of any output objects.\",\n",
    "    agent=wf_runner.agent\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ff2f828-c3f8-4e42-b077-973b75b4a622",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overriding of current TracerProvider is not allowed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mYou are a workflow runner, your role is to efficiently run KBase workflows.\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "    This task involves running multiple apps where the output of one (if any) is fed into the next as input. \n",
      "    Here are the tasks in JSON format: [{\"Step\": 1, \"Name\": \"Quality control\", \"Description\": \"Check reads for quality using FastQC\", \"expect_new_object\": false, \"app_id\": \"kb_fastqc/runFastQC\"}, {\"Step\": 2, \"Name\": \"Reads trimming\", \"Description\": \"Trim low-quality bases and remove adapter sequences from the paired-end reads using Trimmomatic\", \"expect_new_object\": true, \"app_id\": \"kb_trimmomatic/run_trimmomatic\"}, {\"Step\": 3, \"Name\": \"Genome Assembly\", \"Description\": \"Perform de novo assembly of the trimmed reads into contigs and scaffolds using SPAdes.\", \"expect_new_object\": true, \"app_id\": \"kb_SPAdes/run_SPAdes\"}, {\"Step\": 4, \"Name\": \"Assembly Quality Assessment\", \"Description\": \"Obtain assembly statistics such as N50, total length, and number of contigs using QUAST.\", \"expect_new_object\": false, \"app_id\": \"kb_quast/run_QUAST_app\"}, {\"Step\": 5, \"Name\": \"Genome Annotation\", \"Description\": \"Annotate the assembled genome using Prokka to provide insights into gene content and functional elements.\", \"expect_new_object\": true, \"app_id\": \"ProkkaAnnotation/annotate_contigs\"}, {\"Step\": 6, \"Name\": \"Genome Completeness Estimation\", \"Description\": \"Estimate genome completeness and contamination levels using CheckM.\", \"expect_new_object\": false, \"app_id\": \"kb_Msuite/run_checkM_lineage_wf\"}, {\"Step\": 7, \"Name\": \"Build Genome Set\", \"Description\": \"Build a Genome Set to run through GTDB-tk\", \"expect_new_object\": true, \"app_id\": \"kb_SetUtilities/KButil_Build_GenomeSet\"}, {\"Step\": 8, \"Name\": \"Taxonomic Classification\", \"Description\": \"Perform taxonomic classification to confirm organism identity using GTDB-Tk.\", \"app_id\": \"kb_gtdbtk/run_kb_gtdbtk_classify_wf\", \"expect_new_object\": true}].\n",
      "    If any task has \"expect_new_object\" set to True, then that should receive a new data object in its output as a \"created_object\". That object should be used as input for the next task.\n",
      "    If a task as \"expect_new_object\" set to False, then that should not receive a new object to use in the next task. In that case, use the same input object from the previous step for the next one.\n",
      "    These steps must be run sequentially. \n",
      "    These must be run in the narrative with id 211157 and start with using the paired-end reads object 211157/2/1 with name rhodo.art.q20.PE.reads.\n",
      "    If any step ends with an error, immediately stop the task and end with an error.\n",
      "    In the end, return a brief summary of steps taken and resulting output objects.\n",
      "    \u001b[00m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overriding of current TracerProvider is not allowed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            From the given KBase app id, kb_fastqc/runFastQC, fetch the list of parameters needed to run it. Use the App and Job manager agent\n",
      "            for assistance. Using the data object with UPA \"211157/2/1\", populate a dictionary\n",
      "            with the parameters where the keys are parameter ids, and values are the proper parameter values, or their\n",
      "            default values if no value can be found or calculated.\n",
      "            Any input object parameter must be the input object UPA.\n",
      "            Be sure to make sure there is a non-null value for any parameter that is not optional.\n",
      "            Any parameter that has a true value for \"is_output_object\" must have a valid name for the new object. The new object name should be based on\n",
      "            the input object name, not its UPA. If the input object name is not available, the Workspace Manager can assist.\n",
      "            If the parameter type is 'dropdown', use the allowed 'name' option to determine what should be used, but only set the\n",
      "            'value' associated with that name, or the default value if any.\n",
      "            Only alphanumeric characters and underscores are allowed in new object names.\n",
      "            Return the dictionary of inputs, the app id, and the\n",
      "            narrative id 211157 for use in the next task. Do not add comments or other text. If the parameters are rejected, examine the reason why and\n",
      "            reform them. The dictionary of inputs and the app id must not be combined into a single dictionary.\n",
      "            \u001b[00m\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mWhat are the parameters needed to run the KBase app with the id 'kb_fastqc/runFastQC'?\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mget-app-parameters\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"app_id\\\": \\\"kb_fastqc/runFastQC\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "{\"input_file_ref\": {\"id\": \"input_file_ref\", \"ui_name\": \"Read Library/RNA-seq Sample Set\", \"short_hint\": \"Select the Read Library or RNA-seq Sample Set to run quality control on.\", \"is_output_object\": false, \"type\": \"data_object\", \"allowed_data_type\": [\"KBaseFile.PairedEndLibrary\", \"KBaseAssembly.PairedEndLibrary\", \"KBaseAssembly.SingleEndLibrary\", \"KBaseFile.SingleEndLibrary\", \"KBaseRNASeq.RNASeqSampleSet\"], \"multiple\": false, \"optional\": false, \"default_value\": null}}\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\"input_file_ref\": {\"id\": \"input_file_ref\", \"ui_name\": \"Read Library/RNA-seq Sample Set\", \"short_hint\": \"Select the Read Library or RNA-seq Sample Set to run quality control on.\", \"is_output_object\": false, \"type\": \"data_object\", \"allowed_data_type\": [\"KBaseFile.PairedEndLibrary\", \"KBaseAssembly.PairedEndLibrary\", \"KBaseAssembly.SingleEndLibrary\", \"KBaseFile.SingleEndLibrary\", \"KBaseRNASeq.RNASeqSampleSet\"], \"multiple\": false, \"optional\": false, \"default_value\": null}}\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mTo complete the task, I need to first fetch the list of parameters needed to run the KBase app with the id \"kb_fastqc/runFastQC\" using the Job and App Manager. Then, I will need to populate a dictionary with these parameters, ensuring that all non-optional parameters have non-null values and that any output object parameters have valid names. I will also need to ensure that the input object parameter is set to the UPA \"211157/2/1\". \n",
      "First, I will ask the Job and App Manager for the list of parameters for the app \"kb_fastqc/runFastQC\".\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mAsk question to coworker\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"question\\\": \\\"What are the parameters needed to run the KBase app with the id 'kb_fastqc/runFastQC'?\\\", \\\"context\\\": \\\"I need to fetch the list of parameters for the KBase app with the id 'kb_fastqc/runFastQC' to populate a dictionary with the necessary values for running the app. This is part of a task to ensure all non-optional parameters have non-null values and any output object parameters have valid names.\\\", \\\"coworker\\\": \\\"Job and App Manager\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "{\"input_file_ref\": {\"id\": \"input_file_ref\", \"ui_name\": \"Read Library/RNA-seq Sample Set\", \"short_hint\": \"Select the Read Library or RNA-seq Sample Set to run quality control on.\", \"is_output_object\": false, \"type\": \"data_object\", \"allowed_data_type\": [\"KBaseFile.PairedEndLibrary\", \"KBaseAssembly.PairedEndLibrary\", \"KBaseAssembly.SingleEndLibrary\", \"KBaseFile.SingleEndLibrary\", \"KBaseRNASeq.RNASeqSampleSet\"], \"multiple\": false, \"optional\": false, \"default_value\": null}}\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"narrative_id\": 211157,\n",
      "  \"app_id\": \"kb_fastqc/runFastQC\",\n",
      "  \"app_params\": {\n",
      "    \"input_file_ref\": \"211157/2/1\"\n",
      "  }\n",
      "}\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Using the app parameters and app id, use provided tools to start a new KBase app, which will return a job id.\n",
      "            Use that job id to create a new App Cell in the narrative such that narrative_id=211157. Return only the\n",
      "            job id.\n",
      "            \u001b[00m\n",
      "starting start_job tool\n",
      "narrative_id: 211157\n",
      "app_id: kb_fastqc/runFastQC\n",
      "params: {'input_file_ref': '211157/2/1'}\n",
      "starting JobAgent._start_job\n",
      "narrative_id: 211157\n",
      "app_id: kb_fastqc/runFastQC\n",
      "params: {'input_file_ref': '211157/2/1'}\n",
      "{'method': 'kb_fastqc.runFastQC', 'service_ver': '7e67f706dcbfa6008b7b07585fadff3e790d83f0', 'params': [{'input_ws': 'wjriehl:narrative_1742916273434', 'input_file_ref': '211157/2/1'}], 'app_id': 'kb_fastqc/runFastQC', 'wsid': 211157, 'meta': {'cell_id': '5fed6d6f-f652-490c-b14e-d81efe7ccf19', 'run_id': '53e9f03b-7895-4805-8650-c10283ad4c61', 'tag': 'release'}, 'source_ws_objects': ['211157/2/1']}\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mstart-job\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"app_id\\\": \\\"kb_fastqc/runFastQC\\\", \\\"params\\\": {\\\"input_file_ref\\\": \\\"211157/2/1\\\"}}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "67e2f276138d24fd8da1908b\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "67e2f276138d24fd8da1908b\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Using the job id, create an app cell in narrative 211157. The add_app_cell tool is useful here.\n",
      "            \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92madd-app-cell\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"job_id\\\": \\\"67e2f276138d24fd8da1908b\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "success\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "success\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Using the job id, monitor the progress of the running job. The monitor_job tool is helpful here.\n",
      "            If completed, continue to the next task. If in an error state, summarize the error for the user and stop.\n",
      "            \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mThought: I need to monitor the job using the provided job ID to determine its final status.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mmonitor-job\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"job_id\\\": \\\"67e2f276138d24fd8da1908b\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "job_id='67e2f276138d24fd8da1908b' job_status='completed' created_objects=[] job_error=None report_upa='211157/3/1'\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"job_id\": \"67e2f276138d24fd8da1908b\",\n",
      "  \"job_status\": \"completed\",\n",
      "  \"created_objects\": [],\n",
      "  \"job_error\": null,\n",
      "  \"report_upa\": \"211157/3/1\"\n",
      "}\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mGet the report UPA from the job results (a string with format number/number/number), an\n",
      "            use a tool to get the report object from the Workspace service. UPAs are unique permanent addresses used to identify data objects.\n",
      "            The Workspace Manager will be helpful here. Make sure the delegated agent uses a tool for interacting with KBase.\n",
      "            If there was an error in the job results, return a string saying so and summarizing the error.\n",
      "            If no report is available, either because there is no report UPA or the report object is unavailable, return a string saying so.\n",
      "            Otherwise, return the full report text. Your final answer MUST be the report text.\n",
      "            \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mget-report\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"upa\\\": \\\"211157/3/1\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "file 1: rhodo.art.q20.PE.reads_211157_2_1.rev_fastqc.zip:\n",
      "##FastQC\t0.12.1\n",
      ">>Basic Statistics\tpass\n",
      "#Measure\tValue\n",
      "Filename\trhodo.art.q20.PE.reads_211157_2_1.rev.fastq\n",
      "File type\tConventional base calls\n",
      "Encoding\tSanger / Illumina 1.9\n",
      "Total Sequences\t193053\n",
      "Total Bases\t19.3 Mbp\n",
      "Sequences flagged as poor quality\t0\n",
      "Sequence length\t100\n",
      "%GC\t66\n",
      ">>END_MODULE\n",
      ">>Per base sequence quality\tpass\n",
      "#Base\tMean\tMedian\tLower Quartile\tUpper Quartile\t10th Percentile\t90th Percentile\n",
      "1\t48.673524886948144\t51.0\t48.0\t53.0\t36.0\t54.0\n",
      "2\t49.26894686951252\t51.0\t50.0\t53.0\t43.0\t54.0\n",
      "3\t48.860323330898765\t51.0\t50.0\t54.0\t36.0\t54.0\n",
      "4\t48.92431094051892\t55.0\t39.0\t57.0\t30.0\t57.0\n",
      "5\t49.84705236385863\t55.0\t48.0\t57.0\t30.0\t57.0\n",
      "6\t50.67216774668096\t55.0\t50.0\t57.0\t31.0\t57.0\n",
      "7\t50.97477894671412\t55.0\t52.0\t57.0\t31.0\t57.0\n",
      "8\t50.92606175506208\t55.0\t52.0\t57.0\t31.0\t57.0\n",
      "9\t52.17873848114248\t57.0\t52.0\t59.0\t31.0\t59.0\n",
      "10-11\t52.118949200478625\t57.0\t52.0\t59.0\t30.0\t59.0\n",
      "12-13\t51.985366712768\t57.0\t52.0\t59.0\t30.0\t59.0\n",
      "14-15\t52.707072669163395\t58.0\t52.0\t60.0\t30.0\t61.0\n",
      "16-17\t52.567766364677055\t58.0\t51.5\t60.0\t30.0\t61.0\n",
      "18-19\t52.424347199991715\t58.0\t51.0\t60.0\t30.0\t61.0\n",
      "20-21\t52.23753321626704\t58.0\t51.0\t60.0\t29.5\t61.0\n",
      "22-23\t52.08386039066992\t58.0\t50.5\t60.0\t28.5\t61.0\n",
      "24-25\t51.804118558116166\t58.0\t50.0\t60.0\t27.0\t61.0\n",
      "26-27\t51.64548077470953\t58.0\t50.0\t60.0\t23.5\t61.0\n",
      "28-29\t51.45490875562669\t57.5\t50.0\t60.0\t22.0\t61.0\n",
      "30-31\t51.30439309412441\t57.0\t50.0\t60.0\t22.0\t61.0\n",
      "32-33\t51.24059714171756\t57.0\t50.0\t60.0\t22.0\t61.0\n",
      "34-35\t51.27651991940037\t57.0\t50.0\t60.0\t22.0\t61.0\n",
      "36-37\t51.27482608402874\t58.0\t50.0\t60.0\t22.0\t61.0\n",
      "38-39\t51.216018399092476\t58.0\t50.0\t60.0\t22.0\t61.0\n",
      "40-41\t51.093321523105054\t57.5\t50.0\t60.0\t22.0\t61.0\n",
      "42-43\t50.925854558074725\t57.0\t49.0\t60.0\t22.0\t61.0\n",
      "44-45\t50.693620404759315\t57.0\t48.5\t60.0\t22.0\t61.0\n",
      "46-47\t50.48902373959483\t57.0\t48.0\t60.0\t22.0\t61.0\n",
      "48-49\t50.22476470192123\t57.0\t47.0\t60.0\t22.0\t61.0\n",
      "50-51\t49.328837158707714\t55.0\t45.5\t59.0\t22.0\t60.5\n",
      "52-53\t49.38256333752907\t55.5\t46.0\t59.0\t22.0\t60.0\n",
      "54-55\t49.81310054751803\t56.0\t47.0\t59.5\t22.0\t61.0\n",
      "56-57\t49.5891568636592\t56.0\t46.0\t59.5\t22.0\t61.0\n",
      "58-59\t49.33220410975225\t55.5\t46.0\t59.0\t22.0\t61.0\n",
      "60-61\t49.084686588657\t55.0\t46.0\t59.0\t22.0\t61.0\n",
      "62-63\t48.835125587273964\t55.0\t46.0\t59.0\t22.0\t60.5\n",
      "64-65\t48.48487720988537\t55.0\t45.0\t58.5\t22.0\t60.0\n",
      "66-67\t48.11440122660616\t54.5\t44.0\t58.0\t22.0\t60.0\n",
      "68-69\t47.713368867616666\t54.0\t43.0\t57.5\t22.0\t60.0\n",
      "70-71\t47.27160417087536\t54.0\t42.0\t57.0\t22.0\t59.0\n",
      "72-73\t46.8268972769136\t54.0\t41.5\t56.0\t22.0\t59.0\n",
      "74-75\t46.444196671380396\t53.5\t40.5\t56.0\t22.0\t59.0\n",
      "76-77\t46.056647656343074\t53.0\t39.5\t55.0\t22.0\t57.5\n",
      "78-79\t45.66039636783681\t53.0\t38.0\t55.0\t22.0\t57.0\n",
      "80-81\t45.31436185917857\t52.5\t37.5\t55.0\t22.0\t56.5\n",
      "82-83\t44.84410757667584\t52.0\t31.5\t55.0\t22.0\t56.0\n",
      "84-85\t44.50876961248983\t52.0\t27.5\t55.0\t22.0\t56.0\n",
      "86-87\t44.21225259384728\t52.0\t27.0\t55.0\t22.0\t55.0\n",
      "88-89\t43.7673281430488\t51.0\t22.0\t54.5\t22.0\t55.0\n",
      "90-91\t43.423715249180276\t51.0\t22.0\t54.0\t22.0\t55.0\n",
      "92-93\t43.0059336037254\t51.0\t22.0\t54.0\t22.0\t55.0\n",
      "94-95\t42.56390214086287\t51.0\t22.0\t54.0\t22.0\t55.0\n",
      "96-97\t42.07887471316167\t50.5\t22.0\t54.0\t22.0\t55.0\n",
      "98-99\t41.42954007448732\t50.0\t22.0\t54.0\t22.0\t55.0\n",
      "100\t42.28271510932231\t50.0\t26.0\t54.0\t26.0\t55.0\n",
      ">>END_MODULE\n",
      ">>Per sequence quality scores\tpass\n",
      "#Quality\tCount\n",
      "41\t1.0\n",
      "42\t2.0\n",
      "43\t88.0\n",
      "44\t713.0\n",
      "45\t3877.0\n",
      "46\t14516.0\n",
      "47\t36479.0\n",
      "48\t57131.0\n",
      "49\t50333.0\n",
      "50\t23922.0\n",
      "51\t5468.0\n",
      "52\t512.0\n",
      "53\t11.0\n",
      ">>END_MODULE\n",
      ">>Per base sequence content\tpass\n",
      "#Base\tG\tA\tT\tC\n",
      "1\t33.31002367225581\t16.73219271391794\t16.78658192309884\t33.171201690727415\n",
      "2\t33.45091762365775\t16.62393228802453\t16.65552982859629\t33.26962025972142\n",
      "3\t33.31986552915521\t16.60632054409929\t16.66796164783764\t33.405852278907865\n",
      "4\t33.53638638094203\t16.656047821064682\t16.586118837832096\t33.2214469601612\n",
      "5\t33.53431441106846\t16.643616001823332\t16.50168606548461\t33.3203835216236\n",
      "6\t33.268584274784644\t16.812999538986702\t16.567471108970075\t33.35094507725857\n",
      "7\t33.21885699781925\t16.742034570817342\t16.78295597582011\t33.2561524555433\n",
      "8\t33.130280285724645\t16.834237230190674\t16.709401045308798\t33.32608143877588\n",
      "9\t33.40119034669236\t16.57420501105914\t16.7637902544897\t33.2608143877588\n",
      "10-11\t33.268584274784644\t16.76327226202131\t16.616939389701272\t33.351204073492774\n",
      "12-13\t33.279980109089216\t16.700854169580374\t16.71302699258753\t33.30613872874289\n",
      "14-15\t33.39083049732457\t16.64128503571558\t16.635587118563297\t33.33229734839656\n",
      "16-17\t33.29370690950154\t16.66614867419828\t16.729343755341798\t33.31080066095839\n",
      "18-19\t33.26469933127172\t16.704221120624908\t16.707070079201046\t33.32400946890232\n",
      "20-21\t33.3299663822888\t16.809373591707978\t16.67288257628734\t33.18777744971588\n",
      "22-23\t33.19917328402045\t16.711732011416554\t16.836568196298426\t33.252526508264566\n",
      "24-25\t33.35275805089794\t16.676249527331873\t16.724940819360487\t33.2460516024097\n",
      "26-27\t33.42605398517506\t16.765085235660674\t16.63144317881618\t33.17741760034809\n",
      "28-29\t33.272469218297566\t16.76534423189487\t16.708883052840413\t33.25330349696715\n",
      "30-31\t33.26236836516397\t16.71639394363206\t16.739185612241197\t33.28205207896277\n",
      "32-33\t33.2657353162085\t16.69748721853584\t16.65552982859629\t33.38124763665936\n",
      "34-35\t33.228957850952845\t16.66614867419828\t16.747214495501236\t33.35767897934764\n",
      "36-37\t33.24035368525742\t16.693343278788728\t16.655788824830488\t33.41051421112337\n",
      "38-39\t33.3105416647242\t16.73607765743086\t16.7290847591076\t33.22429591873734\n",
      "40-41\t33.31623958187648\t16.6552708323621\t16.602953593054757\t33.42553599270667\n",
      "42-43\t33.3499090923218\t16.70836506037202\t16.667702651603445\t33.27402319570273\n",
      "44-45\t33.32608143877588\t16.75679735616644\t16.669774621477004\t33.24734658358067\n",
      "46-47\t33.28257007143116\t16.566953116501686\t16.778294043604607\t33.372182768462544\n",
      "48-49\t33.18337451373457\t16.739962600943784\t16.76197728085034\t33.314685604471315\n",
      "50-51\t33.45091762365775\t16.689976327744194\t16.685573391762883\t33.17353265683517\n",
      "52-53\t33.22325993380056\t16.627817231537453\t16.807819614302808\t33.34110322035917\n",
      "54-55\t33.373477749633516\t16.73167472144955\t16.628335224005845\t33.26651230491109\n",
      "56-57\t33.22740387354768\t16.699818184643593\t16.702667143219738\t33.37011079858899\n",
      "58-59\t33.25174951956198\t16.80367567455569\t16.65216287755176\t33.29241192833056\n",
      "60-61\t33.19710131414689\t16.77104214904715\t16.747214495501236\t33.28464204130472\n",
      "62-63\t33.35664299441086\t16.68816335410483\t16.661227745748576\t33.29396590573573\n",
      "64-65\t33.24061268149161\t16.739962600943784\t16.564881146628128\t33.45454357093648\n",
      "66-67\t33.2802391053234\t16.754984382527077\t16.73944460847539\t33.225331903674125\n",
      "68-69\t33.33851325801723\t16.695156252428088\t16.676249527331873\t33.29008096222281\n",
      "70-71\t33.36777983248124\t16.677285512268654\t16.666925662900862\t33.28800899234925\n",
      "72-73\t33.338772254251424\t16.725458811828876\t16.701372162048763\t33.23439677187094\n",
      "74-75\t33.36855682118382\t16.65760179846985\t16.697228222301646\t33.27661315804468\n",
      "76-77\t33.26728929361367\t16.66459469679311\t16.76171828461614\t33.306397724977074\n",
      "78-79\t33.22792186601607\t16.763531258255504\t16.659414772109212\t33.34913210361921\n",
      "80-81\t33.405852278907865\t16.66096874951438\t16.68376041812352\t33.249418553454234\n",
      "82-83\t33.25045453839101\t16.720278887144982\t16.720019890910788\t33.309246683553226\n",
      "84-85\t33.26987925595562\t16.795646791295653\t16.74591951433026\t33.188554438418464\n",
      "86-87\t33.25304450073296\t16.788135900504006\t16.736336653665056\t33.22248294509798\n",
      "88-89\t33.20616618234371\t16.78191999088333\t16.811186565347338\t33.20072726142562\n",
      "90-91\t33.17897157775326\t16.69075331644678\t16.83242425655131\t33.29785084924865\n",
      "92-93\t33.297332856780265\t16.67883948967382\t16.756538359932247\t33.26728929361367\n",
      "94-95\t33.341362216593375\t16.652421873785954\t16.704221120624908\t33.30199478899577\n",
      "96-97\t33.2312888170606\t16.712509000119137\t16.664076704324717\t33.392125478495544\n",
      "98-99\t33.31960653292101\t16.773891107623296\t16.621083329448393\t33.285419030007304\n",
      "100\t33.04947346065588\t16.768452186705204\t16.67935748214221\t33.5027168704967\n",
      ">>END_MODULE\n",
      ">>Per sequence GC content\tpass\n",
      "#GC Content\tCount\n",
      "0\t0.0\n",
      "1\t0.0\n",
      "2\t0.0\n",
      "3\t0.0\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      "10\t0.0\n",
      "11\t0.0\n",
      "12\t0.0\n",
      "13\t0.0\n",
      "14\t0.0\n",
      "15\t0.0\n",
      "16\t0.0\n",
      "17\t0.0\n",
      "18\t0.0\n",
      "19\t0.0\n",
      "20\t0.0\n",
      "21\t0.0\n",
      "22\t0.0\n",
      "23\t0.0\n",
      "24\t0.0\n",
      "25\t0.0\n",
      "26\t0.0\n",
      "27\t0.0\n",
      "28\t0.0\n",
      "29\t0.5\n",
      "30\t4.0\n",
      "31\t6.5\n",
      "32\t6.0\n",
      "33\t6.0\n",
      "34\t6.0\n",
      "35\t6.5\n",
      "36\t8.5\n",
      "37\t9.0\n",
      "38\t16.5\n",
      "39\t32.0\n",
      "40\t41.5\n",
      "41\t55.0\n",
      "42\t78.0\n",
      "43\t104.0\n",
      "44\t131.5\n",
      "45\t174.5\n",
      "46\t219.0\n",
      "47\t251.5\n",
      "48\t297.0\n",
      "49\t366.5\n",
      "50\t455.5\n",
      "51\t573.0\n",
      "52\t713.0\n",
      "53\t865.5\n",
      "54\t1093.5\n",
      "55\t1464.0\n",
      "56\t1944.5\n",
      "57\t2577.5\n",
      "58\t3373.5\n",
      "59\t4334.5\n",
      "60\t5514.0\n",
      "61\t6757.5\n",
      "62\t8163.0\n",
      "63\t9629.0\n",
      "64\t10961.5\n",
      "65\t12046.0\n",
      "66\t12786.0\n",
      "67\t13309.0\n",
      "68\t13394.0\n",
      "69\t12948.0\n",
      "70\t12086.5\n",
      "71\t11059.0\n",
      "72\t9890.5\n",
      "73\t8478.0\n",
      "74\t7089.5\n",
      "75\t5679.5\n",
      "76\t4282.5\n",
      "77\t3213.5\n",
      "78\t2329.5\n",
      "79\t1604.0\n",
      "80\t1091.5\n",
      "81\t671.0\n",
      "82\t394.5\n",
      "83\t233.5\n",
      "84\t127.0\n",
      "85\t58.0\n",
      "86\t23.5\n",
      "87\t11.5\n",
      "88\t3.0\n",
      "89\t1.5\n",
      "90\t2.0\n",
      "91\t1.0\n",
      "92\t0.0\n",
      "93\t0.0\n",
      "94\t0.0\n",
      "95\t0.0\n",
      "96\t0.0\n",
      "97\t0.0\n",
      "98\t0.0\n",
      "99\t0.0\n",
      "100\t0.0\n",
      ">>END_MODULE\n",
      ">>Per base N content\tpass\n",
      "#Base\tN-Count\n",
      "1\t0.0\n",
      "2\t0.0\n",
      "3\t0.0\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      "10-11\t0.0\n",
      "12-13\t0.0\n",
      "14-15\t0.0\n",
      "16-17\t0.0\n",
      "18-19\t0.0\n",
      "20-21\t0.0\n",
      "22-23\t0.0\n",
      "24-25\t0.0\n",
      "26-27\t0.0\n",
      "28-29\t0.0\n",
      "30-31\t0.0\n",
      "32-33\t0.0\n",
      "34-35\t0.0\n",
      "36-37\t0.0\n",
      "38-39\t0.0\n",
      "40-41\t0.0\n",
      "42-43\t0.0\n",
      "44-45\t0.0\n",
      "46-47\t0.0\n",
      "48-49\t0.0\n",
      "50-51\t0.0\n",
      "52-53\t0.0\n",
      "54-55\t0.0\n",
      "56-57\t0.0\n",
      "58-59\t0.0\n",
      "60-61\t0.0\n",
      "62-63\t0.0\n",
      "64-65\t0.0\n",
      "66-67\t0.0\n",
      "68-69\t0.0\n",
      "70-71\t0.0\n",
      "72-73\t0.0\n",
      "74-75\t0.0\n",
      "76-77\t0.0\n",
      "78-79\t0.0\n",
      "80-81\t0.0\n",
      "82-83\t0.0\n",
      "84-85\t0.0\n",
      "86-87\t0.0\n",
      "88-89\t0.0\n",
      "90-91\t0.0\n",
      "92-93\t0.0\n",
      "94-95\t0.0\n",
      "96-97\t0.0\n",
      "98-99\t0.0\n",
      "100\t0.0\n",
      ">>END_MODULE\n",
      ">>Sequence Length Distribution\tpass\n",
      "#Length\tCount\n",
      "100\t193053.0\n",
      ">>END_MODULE\n",
      ">>Sequence Duplication Levels\tpass\n",
      "#Total Deduplicated Percentage\t98.5243294134104\n",
      "#Duplication Level\tPercentage of total\n",
      "1\t97.05966508910315\n",
      "2\t2.9073161240498107\n",
      "3\t0.033018786847037844\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      ">10\t0.0\n",
      ">50\t0.0\n",
      ">100\t0.0\n",
      ">500\t0.0\n",
      ">1k\t0.0\n",
      ">5k\t0.0\n",
      ">10k+\t0.0\n",
      ">>END_MODULE\n",
      ">>Overrepresented sequences\tpass\n",
      ">>END_MODULE\n",
      ">>Adapter Content\tpass\n",
      "#Position\tIllumina Universal Adapter\tIllumina Small RNA 3' Adapter\tIllumina Small RNA 5' Adapter\tNextera Transposase Sequence\tPolyA\tPolyG\n",
      "1\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "2\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "3\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "4\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "5\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "6\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "7\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "8\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "9\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "10-11\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "12-13\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "14-15\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "16-17\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "18-19\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "20-21\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "22-23\t0.0\t5.179924683895096E-4\t0.0\t0.0\t0.0\t0.0\n",
      "24-25\t0.0\t0.0010359849367790191\t0.0\t0.0\t0.0\t0.0\n",
      "26-27\t0.0\t0.0010359849367790191\t0.0\t0.0\t0.0\t0.0\n",
      "28-29\t0.0\t0.0010359849367790191\t0.0\t0.0\t0.0\t0.0\n",
      "30-31\t0.0\t0.0010359849367790191\t0.0\t0.0\t0.0\t0.0\n",
      "32-33\t0.0\t0.001294981170973774\t0.0\t0.0\t0.0\t0.0\n",
      "34-35\t0.0\t0.001553977405168529\t0.0\t0.0\t0.0\t0.0\n",
      "36-37\t0.0\t0.001553977405168529\t0.0\t0.0\t0.0\t0.0\n",
      "38-39\t0.0\t0.001553977405168529\t0.0\t0.0\t0.0\t0.0\n",
      "40-41\t0.0\t0.001553977405168529\t0.0\t0.0\t0.0\t0.0\n",
      "42-43\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "44-45\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "46-47\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "48-49\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "50-51\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "52-53\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "54-55\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "56-57\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "58-59\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "60-61\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "62-63\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "64-65\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "66-67\t0.0\t0.0033669510445318127\t0.0\t0.0\t0.0\t0.0\n",
      "68-69\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "70-71\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "72-73\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "74-75\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "76-77\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "78-79\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "80-81\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "82-83\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "84-85\t0.0\t0.003884943512921322\t0.0\t0.0\t0.0\t0.0\n",
      "86-87\t0.0\t0.004661932215505586\t0.0\t0.0\t0.0\t0.0\n",
      "88-89\t0.0\t0.005179924683895096\t0.0\t0.0\t0.0\t0.0\n",
      ">>END_MODULE\n",
      "\n",
      "file 2: rhodo.art.q20.PE.reads_211157_2_1.fwd_fastqc.zip:\n",
      "##FastQC\t0.12.1\n",
      ">>Basic Statistics\tpass\n",
      "#Measure\tValue\n",
      "Filename\trhodo.art.q20.PE.reads_211157_2_1.fwd.fastq\n",
      "File type\tConventional base calls\n",
      "Encoding\tSanger / Illumina 1.9\n",
      "Total Sequences\t193053\n",
      "Total Bases\t19.3 Mbp\n",
      "Sequences flagged as poor quality\t0\n",
      "Sequence length\t100\n",
      "%GC\t66\n",
      ">>END_MODULE\n",
      ">>Per base sequence quality\tpass\n",
      "#Base\tMean\tMedian\tLower Quartile\tUpper Quartile\t10th Percentile\t90th Percentile\n",
      "1\t50.42625600223773\t51.0\t51.0\t54.0\t47.0\t54.0\n",
      "2\t50.52894800909595\t51.0\t51.0\t54.0\t46.0\t54.0\n",
      "3\t50.32879053938556\t51.0\t51.0\t54.0\t46.0\t54.0\n",
      "4\t53.89255800220665\t57.0\t55.0\t57.0\t48.0\t57.0\n",
      "5\t53.676254707256554\t57.0\t55.0\t57.0\t48.0\t57.0\n",
      "6\t53.67517210299762\t57.0\t55.0\t57.0\t48.0\t57.0\n",
      "7\t53.71687049670298\t57.0\t55.0\t57.0\t49.0\t57.0\n",
      "8\t53.632339305786495\t57.0\t55.0\t57.0\t48.0\t57.0\n",
      "9\t55.05453942699673\t59.0\t55.0\t59.0\t48.0\t59.0\n",
      "10-11\t54.97349432539251\t59.0\t55.0\t59.0\t48.0\t59.0\n",
      "12-13\t54.85529880395539\t59.0\t55.0\t59.0\t47.0\t59.0\n",
      "14-15\t55.89482680921819\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "16-17\t55.727287325242294\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "18-19\t55.70098107773513\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "20-21\t55.642559297187816\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "22-23\t55.59338109223892\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "24-25\t55.55280156226529\t59.5\t56.0\t61.0\t46.0\t61.0\n",
      "26-27\t55.33956219276572\t59.0\t55.5\t61.0\t45.0\t61.0\n",
      "28-29\t55.19439221353721\t59.0\t55.0\t61.0\t45.0\t61.0\n",
      "30-31\t55.047989412233946\t59.0\t55.0\t61.0\t44.0\t61.0\n",
      "32-33\t54.86606527741087\t59.0\t54.5\t61.0\t44.0\t61.0\n",
      "34-35\t54.67198903927937\t59.0\t54.0\t60.5\t43.5\t61.0\n",
      "36-37\t54.489979435699\t59.0\t54.0\t60.0\t42.0\t61.0\n",
      "38-39\t54.3263663346335\t58.0\t54.0\t60.0\t41.0\t61.0\n",
      "40-41\t54.17223767566419\t58.0\t53.5\t60.0\t39.5\t61.0\n",
      "42-43\t53.97476340694006\t58.0\t53.0\t60.0\t38.5\t61.0\n",
      "44-45\t53.80058325951941\t58.0\t53.0\t60.0\t37.5\t61.0\n",
      "46-47\t53.67100485358942\t58.0\t53.0\t60.0\t36.5\t61.0\n",
      "48-49\t53.44004496174625\t58.0\t53.0\t60.0\t34.5\t61.0\n",
      "50-51\t53.32651655244933\t58.0\t53.0\t60.0\t34.0\t61.0\n",
      "52-53\t53.12265284662761\t58.0\t52.5\t60.0\t32.5\t61.0\n",
      "54-55\t52.65927232418041\t57.0\t52.0\t60.0\t28.0\t61.0\n",
      "56-57\t52.32782707339435\t57.0\t51.0\t60.0\t27.5\t61.0\n",
      "58-59\t52.01091410130897\t56.5\t51.0\t60.0\t26.5\t61.0\n",
      "60-61\t51.797493434445464\t56.0\t51.0\t59.0\t24.0\t61.0\n",
      "62-63\t51.44003201193455\t56.0\t50.5\t59.0\t22.0\t60.5\n",
      "64-65\t51.041892640881\t55.0\t50.0\t59.0\t22.0\t60.0\n",
      "66-67\t50.49370380154672\t55.0\t49.0\t58.5\t22.0\t60.0\n",
      "68-69\t49.959581047691564\t54.5\t48.5\t58.0\t22.0\t60.0\n",
      "70-71\t49.483424241011534\t54.0\t48.0\t57.0\t22.0\t59.5\n",
      "72-73\t48.99674959726086\t54.0\t47.5\t56.5\t22.0\t59.0\n",
      "74-75\t48.62221514299182\t54.0\t46.5\t56.0\t22.0\t59.0\n",
      "76-77\t47.58507508300829\t52.5\t46.0\t55.0\t22.0\t57.0\n",
      "78-79\t48.239677705086166\t54.0\t47.0\t55.0\t22.0\t57.0\n",
      "80-81\t48.06770938550554\t54.0\t47.0\t55.0\t22.0\t57.0\n",
      "82-83\t47.82851341341497\t54.0\t46.5\t55.0\t22.0\t56.5\n",
      "84-85\t47.44304672810057\t53.0\t46.0\t55.0\t22.0\t56.0\n",
      "86-87\t47.02018616649314\t53.0\t45.5\t55.0\t22.0\t56.0\n",
      "88-89\t46.609412959135575\t53.0\t45.0\t55.0\t22.0\t55.0\n",
      "90-91\t46.146970003056154\t52.5\t44.5\t55.0\t22.0\t55.0\n",
      "92-93\t45.50171455507037\t52.0\t41.5\t54.5\t22.0\t55.0\n",
      "94-95\t44.92707443033778\t52.0\t38.5\t54.0\t22.0\t55.0\n",
      "96-97\t44.322284553982584\t52.0\t28.5\t54.0\t22.0\t55.0\n",
      "98-99\t43.466910641119284\t51.0\t22.0\t54.0\t22.0\t55.0\n",
      "100\t45.04022729509513\t51.0\t31.0\t54.0\t31.0\t55.0\n",
      ">>END_MODULE\n",
      ">>Per sequence quality scores\tpass\n",
      "#Quality\tCount\n",
      "45\t3.0\n",
      "46\t43.0\n",
      "47\t287.0\n",
      "48\t2362.0\n",
      "49\t12266.0\n",
      "50\t39684.0\n",
      "51\t67212.0\n",
      "52\t53493.0\n",
      "53\t16149.0\n",
      "54\t1531.0\n",
      "55\t23.0\n",
      ">>END_MODULE\n",
      ">>Per base sequence content\tpass\n",
      "#Base\tG\tA\tT\tC\n",
      "1\t33.40481629397109\t16.693343278788728\t16.566435124033298\t33.33540530320689\n",
      "2\t33.2312888170606\t16.832165260317115\t16.65967376834341\t33.27687215427888\n",
      "3\t33.34213920529596\t16.65656581353307\t16.731156728981162\t33.27013825218981\n",
      "4\t33.26703029737948\t16.737890631070222\t16.652421873785954\t33.342657197764346\n",
      "5\t33.51307671986449\t16.648277934038838\t16.637400092202657\t33.20124525389401\n",
      "6\t33.19502934427334\t16.812999538986702\t16.709401045308798\t33.28257007143116\n",
      "7\t33.35405303206891\t16.74462453315929\t16.65552982859629\t33.245792606175506\n",
      "8\t33.27842613168405\t16.77259612645232\t16.823359388354493\t33.12561835350914\n",
      "9\t33.19451135180495\t16.77415010385749\t16.725976804297265\t33.305361740040304\n",
      "10-11\t33.26521732374012\t16.67832149720543\t16.689976327744194\t33.36648485131026\n",
      "12-13\t33.26185037269558\t16.807560618068614\t16.774927092560073\t33.15566191667573\n",
      "14-15\t33.418284098149215\t16.710437030245583\t16.651903881317565\t33.21937499028764\n",
      "16-17\t33.36104593039217\t16.557370255836478\t16.759387318508388\t33.32219649526296\n",
      "18-19\t33.3149446007055\t16.734782676259886\t16.619788348277417\t33.33048437475719\n",
      "20-21\t33.32737641994685\t16.784509953225278\t16.621342325682583\t33.26677130114528\n",
      "22-23\t33.35819697181603\t16.585341849129513\t16.838122173703592\t33.21833900535086\n",
      "24-25\t33.252526508264566\t16.723645838189512\t16.674695549926703\t33.34913210361921\n",
      "26-27\t33.40248532786333\t16.735300668728275\t16.53121163618281\t33.33100236722558\n",
      "28-29\t33.24786457604906\t16.66588967796408\t16.673659564989926\t33.412586180996925\n",
      "30-31\t33.28593702247569\t16.695415248662286\t16.614349427359326\t33.40429830150269\n",
      "32-33\t33.32245549149715\t16.693343278788728\t16.5858598415979\t33.398341388116215\n",
      "34-35\t33.322714487731346\t16.546751410234496\t16.7886538929724\t33.341880209061756\n",
      "36-37\t33.14323009743438\t16.65656581353307\t16.743847544456703\t33.45635654457585\n",
      "38-39\t33.317016570579064\t16.654234847425318\t16.71354498505592\t33.3152035969397\n",
      "40-41\t33.360009945455396\t16.704221120624908\t16.6651126892615\t33.270656244658205\n",
      "42-43\t33.27195122582918\t16.68764536163644\t16.619270355809025\t33.42113305672535\n",
      "44-45\t33.34680113751146\t16.659414772109212\t16.72519981559468\t33.268584274784644\n",
      "46-47\t33.28645501494408\t16.68971733151\t16.69127130891517\t33.332556344630746\n",
      "48-49\t33.30613872874289\t16.737372638601837\t16.64775994157045\t33.30872869108483\n",
      "50-51\t33.338772254251424\t16.831647267848727\t16.70007718087779\t33.12950329702206\n",
      "52-53\t33.21859800158506\t16.703703128156516\t16.73607765743086\t33.34162121282757\n",
      "54-55\t33.19995027272304\t16.75731534863483\t16.68971733151\t33.353017047132134\n",
      "56-57\t33.26469933127172\t16.694379263725505\t16.810927569113147\t33.22999383588963\n",
      "58-59\t33.33618229190948\t16.668997632774417\t16.709401045308798\t33.285419030007304\n",
      "60-61\t33.16058284512543\t16.75679735616644\t16.72778977793663\t33.354830020771495\n",
      "62-63\t33.1768996078797\t16.740480593412173\t16.737372638601837\t33.34524716010629\n",
      "64-65\t33.26469933127172\t16.728307770405017\t16.743847544456703\t33.263145353866555\n",
      "66-67\t33.27687215427888\t16.653457858722735\t16.69075331644678\t33.37891667055161\n",
      "68-69\t33.24786457604906\t16.72519981559468\t16.731415725215356\t33.2955198831409\n",
      "70-71\t33.354830020771495\t16.672105587584756\t16.72571780806307\t33.24734658358067\n",
      "72-73\t33.32297348396555\t16.63144317881618\t16.81481251262607\t33.23077082459221\n",
      "74-75\t33.3253044500733\t16.741257582114756\t16.660709753280187\t33.27272821453176\n",
      "76-77\t33.30691571744547\t16.77958902477558\t16.65708380600146\t33.25641145177749\n",
      "78-79\t33.304066758869325\t16.58896779640824\t16.82646734316483\t33.280498101557605\n",
      "80-81\t33.31364961953453\t16.735300668728275\t16.616421397232884\t33.33462831450431\n",
      "82-83\t33.29888683418543\t16.718983905974007\t16.632479163752958\t33.3496500960876\n",
      "84-85\t33.36700284377866\t16.76223627708453\t16.64439299052592\t33.226367888610895\n",
      "86-87\t33.240094689023216\t16.696710229833258\t16.732710706386328\t33.33048437475719\n",
      "88-89\t33.33721827684625\t16.71639394363206\t16.74462453315929\t33.2017632463624\n",
      "90-91\t33.19218038569719\t16.752135423950936\t16.692825286320335\t33.36285890403153\n",
      "92-93\t33.41673012074404\t16.723127845721123\t16.554780293494535\t33.305361740040304\n",
      "94-95\t33.26262736139817\t16.808078610537002\t16.617716378403856\t33.311577649660975\n",
      "96-97\t33.268066282316255\t16.66200473445116\t16.764826239426476\t33.3051027438061\n",
      "98-99\t33.15307195433378\t16.767934194236815\t16.752135423950936\t33.32685842747846\n",
      "100\t33.24268465136517\t16.666407670432474\t16.693343278788728\t33.39756439941363\n",
      ">>END_MODULE\n",
      ">>Per sequence GC content\tpass\n",
      "#GC Content\tCount\n",
      "0\t0.0\n",
      "1\t0.0\n",
      "2\t0.0\n",
      "3\t0.0\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      "10\t0.0\n",
      "11\t0.0\n",
      "12\t0.0\n",
      "13\t0.0\n",
      "14\t0.0\n",
      "15\t0.0\n",
      "16\t0.0\n",
      "17\t0.0\n",
      "18\t0.0\n",
      "19\t0.0\n",
      "20\t0.0\n",
      "21\t0.0\n",
      "22\t0.0\n",
      "23\t0.0\n",
      "24\t0.0\n",
      "25\t0.0\n",
      "26\t0.0\n",
      "27\t0.0\n",
      "28\t0.0\n",
      "29\t0.0\n",
      "30\t2.0\n",
      "31\t7.0\n",
      "32\t10.5\n",
      "33\t8.5\n",
      "34\t7.5\n",
      "35\t8.0\n",
      "36\t7.0\n",
      "37\t11.5\n",
      "38\t20.0\n",
      "39\t25.0\n",
      "40\t44.5\n",
      "41\t67.0\n",
      "42\t80.0\n",
      "43\t91.5\n",
      "44\t118.0\n",
      "45\t168.0\n",
      "46\t214.5\n",
      "47\t268.0\n",
      "48\t321.0\n",
      "49\t377.0\n",
      "50\t460.0\n",
      "51\t559.0\n",
      "52\t676.5\n",
      "53\t848.5\n",
      "54\t1064.0\n",
      "55\t1423.0\n",
      "56\t1910.5\n",
      "57\t2505.5\n",
      "58\t3366.0\n",
      "59\t4353.0\n",
      "60\t5474.5\n",
      "61\t6780.0\n",
      "62\t8181.5\n",
      "63\t9598.0\n",
      "64\t10884.5\n",
      "65\t12022.5\n",
      "66\t12985.5\n",
      "67\t13463.0\n",
      "68\t13348.5\n",
      "69\t12959.5\n",
      "70\t12207.5\n",
      "71\t11147.5\n",
      "72\t9880.0\n",
      "73\t8448.5\n",
      "74\t7001.0\n",
      "75\t5576.5\n",
      "76\t4333.0\n",
      "77\t3236.5\n",
      "78\t2315.5\n",
      "79\t1602.0\n",
      "80\t1050.5\n",
      "81\t663.5\n",
      "82\t398.5\n",
      "83\t230.5\n",
      "84\t126.0\n",
      "85\t66.0\n",
      "86\t34.0\n",
      "87\t11.0\n",
      "88\t3.5\n",
      "89\t0.5\n",
      "90\t0.0\n",
      "91\t0.5\n",
      "92\t0.5\n",
      "93\t0.0\n",
      "94\t0.0\n",
      "95\t0.0\n",
      "96\t0.0\n",
      "97\t0.0\n",
      "98\t0.0\n",
      "99\t0.0\n",
      "100\t0.0\n",
      ">>END_MODULE\n",
      ">>Per base N content\tpass\n",
      "#Base\tN-Count\n",
      "1\t0.0\n",
      "2\t0.0\n",
      "3\t0.0\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      "10-11\t0.0\n",
      "12-13\t0.0\n",
      "14-15\t0.0\n",
      "16-17\t0.0\n",
      "18-19\t0.0\n",
      "20-21\t0.0\n",
      "22-23\t0.0\n",
      "24-25\t0.0\n",
      "26-27\t0.0\n",
      "28-29\t0.0\n",
      "30-31\t0.0\n",
      "32-33\t0.0\n",
      "34-35\t0.0\n",
      "36-37\t0.0\n",
      "38-39\t0.0\n",
      "40-41\t0.0\n",
      "42-43\t0.0\n",
      "44-45\t0.0\n",
      "46-47\t0.0\n",
      "48-49\t0.0\n",
      "50-51\t0.0\n",
      "52-53\t0.0\n",
      "54-55\t0.0\n",
      "56-57\t0.0\n",
      "58-59\t0.0\n",
      "60-61\t0.0\n",
      "62-63\t0.0\n",
      "64-65\t0.0\n",
      "66-67\t0.0\n",
      "68-69\t0.0\n",
      "70-71\t0.0\n",
      "72-73\t0.0\n",
      "74-75\t0.0\n",
      "76-77\t0.0\n",
      "78-79\t0.0\n",
      "80-81\t0.0\n",
      "82-83\t0.0\n",
      "84-85\t0.0\n",
      "86-87\t0.0\n",
      "88-89\t0.0\n",
      "90-91\t0.0\n",
      "92-93\t0.0\n",
      "94-95\t0.0\n",
      "96-97\t0.0\n",
      "98-99\t0.0\n",
      "100\t0.0\n",
      ">>END_MODULE\n",
      ">>Sequence Length Distribution\tpass\n",
      "#Length\tCount\n",
      "100\t193053.0\n",
      ">>END_MODULE\n",
      ">>Sequence Duplication Levels\tpass\n",
      "#Total Deduplicated Percentage\t98.51099750971424\n",
      "#Duplication Level\tPercentage of total\n",
      "1\t97.02969767266319\n",
      "2\t2.947194367632699\n",
      "3\t0.023107959704111326\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      ">10\t0.0\n",
      ">50\t0.0\n",
      ">100\t0.0\n",
      ">500\t0.0\n",
      ">1k\t0.0\n",
      ">5k\t0.0\n",
      ">10k+\t0.0\n",
      ">>END_MODULE\n",
      ">>Overrepresented sequences\tpass\n",
      ">>END_MODULE\n",
      ">>Adapter Content\tpass\n",
      "#Position\tIllumina Universal Adapter\tIllumina Small RNA 3' Adapter\tIllumina Small RNA 5' Adapter\tNextera Transposase Sequence\tPolyA\tPolyG\n",
      "1\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "2\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "3\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "4\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "5\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "6\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "7\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "8\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "9\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "10-11\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "12-13\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "14-15\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "16-17\t0.0\t2.589962341947548E-4\t0.0\t0.0\t0.0\t0.0\n",
      "18-19\t0.0\t7.769887025842644E-4\t0.0\t0.0\t0.0\t0.0\n",
      "20-21\t0.0\t0.001553977405168529\t0.0\t0.0\t0.0\t0.0\n",
      "22-23\t0.0\t0.0018129736393632835\t0.0\t0.0\t0.0\t0.0\n",
      "24-25\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "26-27\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "28-29\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "30-31\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "32-33\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "34-35\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "36-37\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "38-39\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "40-41\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "42-43\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "44-45\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "46-47\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "48-49\t0.0\t0.0028489585761423027\t0.0\t0.0\t0.0\t0.0\n",
      "50-51\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "52-53\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "54-55\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "56-57\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "58-59\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "60-61\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "62-63\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "64-65\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "66-67\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "68-69\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "70-71\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "72-73\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "74-75\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "76-77\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "78-79\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "80-81\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "82-83\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "84-85\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "86-87\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "88-89\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      ">>END_MODULE\n",
      "\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "file 1: rhodo.art.q20.PE.reads_211157_2_1.rev_fastqc.zip:\n",
      "##FastQC\t0.12.1\n",
      ">>Basic Statistics\tpass\n",
      "#Measure\tValue\n",
      "Filename\trhodo.art.q20.PE.reads_211157_2_1.rev.fastq\n",
      "File type\tConventional base calls\n",
      "Encoding\tSanger / Illumina 1.9\n",
      "Total Sequences\t193053\n",
      "Total Bases\t19.3 Mbp\n",
      "Sequences flagged as poor quality\t0\n",
      "Sequence length\t100\n",
      "%GC\t66\n",
      ">>END_MODULE\n",
      ">>Per base sequence quality\tpass\n",
      "#Base\tMean\tMedian\tLower Quartile\tUpper Quartile\t10th Percentile\t90th Percentile\n",
      "1\t48.673524886948144\t51.0\t48.0\t53.0\t36.0\t54.0\n",
      "2\t49.26894686951252\t51.0\t50.0\t53.0\t43.0\t54.0\n",
      "3\t48.860323330898765\t51.0\t50.0\t54.0\t36.0\t54.0\n",
      "4\t48.92431094051892\t55.0\t39.0\t57.0\t30.0\t57.0\n",
      "5\t49.84705236385863\t55.0\t48.0\t57.0\t30.0\t57.0\n",
      "6\t50.67216774668096\t55.0\t50.0\t57.0\t31.0\t57.0\n",
      "7\t50.97477894671412\t55.0\t52.0\t57.0\t31.0\t57.0\n",
      "8\t50.92606175506208\t55.0\t52.0\t57.0\t31.0\t57.0\n",
      "9\t52.17873848114248\t57.0\t52.0\t59.0\t31.0\t59.0\n",
      "10-11\t52.118949200478625\t57.0\t52.0\t59.0\t30.0\t59.0\n",
      "12-13\t51.985366712768\t57.0\t52.0\t59.0\t30.0\t59.0\n",
      "14-15\t52.707072669163395\t58.0\t52.0\t60.0\t30.0\t61.0\n",
      "16-17\t52.567766364677055\t58.0\t51.5\t60.0\t30.0\t61.0\n",
      "18-19\t52.424347199991715\t58.0\t51.0\t60.0\t30.0\t61.0\n",
      "20-21\t52.23753321626704\t58.0\t51.0\t60.0\t29.5\t61.0\n",
      "22-23\t52.08386039066992\t58.0\t50.5\t60.0\t28.5\t61.0\n",
      "24-25\t51.804118558116166\t58.0\t50.0\t60.0\t27.0\t61.0\n",
      "26-27\t51.64548077470953\t58.0\t50.0\t60.0\t23.5\t61.0\n",
      "28-29\t51.45490875562669\t57.5\t50.0\t60.0\t22.0\t61.0\n",
      "30-31\t51.30439309412441\t57.0\t50.0\t60.0\t22.0\t61.0\n",
      "32-33\t51.24059714171756\t57.0\t50.0\t60.0\t22.0\t61.0\n",
      "34-35\t51.27651991940037\t57.0\t50.0\t60.0\t22.0\t61.0\n",
      "36-37\t51.27482608402874\t58.0\t50.0\t60.0\t22.0\t61.0\n",
      "38-39\t51.216018399092476\t58.0\t50.0\t60.0\t22.0\t61.0\n",
      "40-41\t51.093321523105054\t57.5\t50.0\t60.0\t22.0\t61.0\n",
      "42-43\t50.925854558074725\t57.0\t49.0\t60.0\t22.0\t61.0\n",
      "44-45\t50.693620404759315\t57.0\t48.5\t60.0\t22.0\t61.0\n",
      "46-47\t50.48902373959483\t57.0\t48.0\t60.0\t22.0\t61.0\n",
      "48-49\t50.22476470192123\t57.0\t47.0\t60.0\t22.0\t61.0\n",
      "50-51\t49.328837158707714\t55.0\t45.5\t59.0\t22.0\t60.5\n",
      "52-53\t49.38256333752907\t55.5\t46.0\t59.0\t22.0\t60.0\n",
      "54-55\t49.81310054751803\t56.0\t47.0\t59.5\t22.0\t61.0\n",
      "56-57\t49.5891568636592\t56.0\t46.0\t59.5\t22.0\t61.0\n",
      "58-59\t49.33220410975225\t55.5\t46.0\t59.0\t22.0\t61.0\n",
      "60-61\t49.084686588657\t55.0\t46.0\t59.0\t22.0\t61.0\n",
      "62-63\t48.835125587273964\t55.0\t46.0\t59.0\t22.0\t60.5\n",
      "64-65\t48.48487720988537\t55.0\t45.0\t58.5\t22.0\t60.0\n",
      "66-67\t48.11440122660616\t54.5\t44.0\t58.0\t22.0\t60.0\n",
      "68-69\t47.713368867616666\t54.0\t43.0\t57.5\t22.0\t60.0\n",
      "70-71\t47.27160417087536\t54.0\t42.0\t57.0\t22.0\t59.0\n",
      "72-73\t46.8268972769136\t54.0\t41.5\t56.0\t22.0\t59.0\n",
      "74-75\t46.444196671380396\t53.5\t40.5\t56.0\t22.0\t59.0\n",
      "76-77\t46.056647656343074\t53.0\t39.5\t55.0\t22.0\t57.5\n",
      "78-79\t45.66039636783681\t53.0\t38.0\t55.0\t22.0\t57.0\n",
      "80-81\t45.31436185917857\t52.5\t37.5\t55.0\t22.0\t56.5\n",
      "82-83\t44.84410757667584\t52.0\t31.5\t55.0\t22.0\t56.0\n",
      "84-85\t44.50876961248983\t52.0\t27.5\t55.0\t22.0\t56.0\n",
      "86-87\t44.21225259384728\t52.0\t27.0\t55.0\t22.0\t55.0\n",
      "88-89\t43.7673281430488\t51.0\t22.0\t54.5\t22.0\t55.0\n",
      "90-91\t43.423715249180276\t51.0\t22.0\t54.0\t22.0\t55.0\n",
      "92-93\t43.0059336037254\t51.0\t22.0\t54.0\t22.0\t55.0\n",
      "94-95\t42.56390214086287\t51.0\t22.0\t54.0\t22.0\t55.0\n",
      "96-97\t42.07887471316167\t50.5\t22.0\t54.0\t22.0\t55.0\n",
      "98-99\t41.42954007448732\t50.0\t22.0\t54.0\t22.0\t55.0\n",
      "100\t42.28271510932231\t50.0\t26.0\t54.0\t26.0\t55.0\n",
      ">>END_MODULE\n",
      ">>Per sequence quality scores\tpass\n",
      "#Quality\tCount\n",
      "41\t1.0\n",
      "42\t2.0\n",
      "43\t88.0\n",
      "44\t713.0\n",
      "45\t3877.0\n",
      "46\t14516.0\n",
      "47\t36479.0\n",
      "48\t57131.0\n",
      "49\t50333.0\n",
      "50\t23922.0\n",
      "51\t5468.0\n",
      "52\t512.0\n",
      "53\t11.0\n",
      ">>END_MODULE\n",
      ">>Per base sequence content\tpass\n",
      "#Base\tG\tA\tT\tC\n",
      "1\t33.31002367225581\t16.73219271391794\t16.78658192309884\t33.171201690727415\n",
      "2\t33.45091762365775\t16.62393228802453\t16.65552982859629\t33.26962025972142\n",
      "3\t33.31986552915521\t16.60632054409929\t16.66796164783764\t33.405852278907865\n",
      "4\t33.53638638094203\t16.656047821064682\t16.586118837832096\t33.2214469601612\n",
      "5\t33.53431441106846\t16.643616001823332\t16.50168606548461\t33.3203835216236\n",
      "6\t33.268584274784644\t16.812999538986702\t16.567471108970075\t33.35094507725857\n",
      "7\t33.21885699781925\t16.742034570817342\t16.78295597582011\t33.2561524555433\n",
      "8\t33.130280285724645\t16.834237230190674\t16.709401045308798\t33.32608143877588\n",
      "9\t33.40119034669236\t16.57420501105914\t16.7637902544897\t33.2608143877588\n",
      "10-11\t33.268584274784644\t16.76327226202131\t16.616939389701272\t33.351204073492774\n",
      "12-13\t33.279980109089216\t16.700854169580374\t16.71302699258753\t33.30613872874289\n",
      "14-15\t33.39083049732457\t16.64128503571558\t16.635587118563297\t33.33229734839656\n",
      "16-17\t33.29370690950154\t16.66614867419828\t16.729343755341798\t33.31080066095839\n",
      "18-19\t33.26469933127172\t16.704221120624908\t16.707070079201046\t33.32400946890232\n",
      "20-21\t33.3299663822888\t16.809373591707978\t16.67288257628734\t33.18777744971588\n",
      "22-23\t33.19917328402045\t16.711732011416554\t16.836568196298426\t33.252526508264566\n",
      "24-25\t33.35275805089794\t16.676249527331873\t16.724940819360487\t33.2460516024097\n",
      "26-27\t33.42605398517506\t16.765085235660674\t16.63144317881618\t33.17741760034809\n",
      "28-29\t33.272469218297566\t16.76534423189487\t16.708883052840413\t33.25330349696715\n",
      "30-31\t33.26236836516397\t16.71639394363206\t16.739185612241197\t33.28205207896277\n",
      "32-33\t33.2657353162085\t16.69748721853584\t16.65552982859629\t33.38124763665936\n",
      "34-35\t33.228957850952845\t16.66614867419828\t16.747214495501236\t33.35767897934764\n",
      "36-37\t33.24035368525742\t16.693343278788728\t16.655788824830488\t33.41051421112337\n",
      "38-39\t33.3105416647242\t16.73607765743086\t16.7290847591076\t33.22429591873734\n",
      "40-41\t33.31623958187648\t16.6552708323621\t16.602953593054757\t33.42553599270667\n",
      "42-43\t33.3499090923218\t16.70836506037202\t16.667702651603445\t33.27402319570273\n",
      "44-45\t33.32608143877588\t16.75679735616644\t16.669774621477004\t33.24734658358067\n",
      "46-47\t33.28257007143116\t16.566953116501686\t16.778294043604607\t33.372182768462544\n",
      "48-49\t33.18337451373457\t16.739962600943784\t16.76197728085034\t33.314685604471315\n",
      "50-51\t33.45091762365775\t16.689976327744194\t16.685573391762883\t33.17353265683517\n",
      "52-53\t33.22325993380056\t16.627817231537453\t16.807819614302808\t33.34110322035917\n",
      "54-55\t33.373477749633516\t16.73167472144955\t16.628335224005845\t33.26651230491109\n",
      "56-57\t33.22740387354768\t16.699818184643593\t16.702667143219738\t33.37011079858899\n",
      "58-59\t33.25174951956198\t16.80367567455569\t16.65216287755176\t33.29241192833056\n",
      "60-61\t33.19710131414689\t16.77104214904715\t16.747214495501236\t33.28464204130472\n",
      "62-63\t33.35664299441086\t16.68816335410483\t16.661227745748576\t33.29396590573573\n",
      "64-65\t33.24061268149161\t16.739962600943784\t16.564881146628128\t33.45454357093648\n",
      "66-67\t33.2802391053234\t16.754984382527077\t16.73944460847539\t33.225331903674125\n",
      "68-69\t33.33851325801723\t16.695156252428088\t16.676249527331873\t33.29008096222281\n",
      "70-71\t33.36777983248124\t16.677285512268654\t16.666925662900862\t33.28800899234925\n",
      "72-73\t33.338772254251424\t16.725458811828876\t16.701372162048763\t33.23439677187094\n",
      "74-75\t33.36855682118382\t16.65760179846985\t16.697228222301646\t33.27661315804468\n",
      "76-77\t33.26728929361367\t16.66459469679311\t16.76171828461614\t33.306397724977074\n",
      "78-79\t33.22792186601607\t16.763531258255504\t16.659414772109212\t33.34913210361921\n",
      "80-81\t33.405852278907865\t16.66096874951438\t16.68376041812352\t33.249418553454234\n",
      "82-83\t33.25045453839101\t16.720278887144982\t16.720019890910788\t33.309246683553226\n",
      "84-85\t33.26987925595562\t16.795646791295653\t16.74591951433026\t33.188554438418464\n",
      "86-87\t33.25304450073296\t16.788135900504006\t16.736336653665056\t33.22248294509798\n",
      "88-89\t33.20616618234371\t16.78191999088333\t16.811186565347338\t33.20072726142562\n",
      "90-91\t33.17897157775326\t16.69075331644678\t16.83242425655131\t33.29785084924865\n",
      "92-93\t33.297332856780265\t16.67883948967382\t16.756538359932247\t33.26728929361367\n",
      "94-95\t33.341362216593375\t16.652421873785954\t16.704221120624908\t33.30199478899577\n",
      "96-97\t33.2312888170606\t16.712509000119137\t16.664076704324717\t33.392125478495544\n",
      "98-99\t33.31960653292101\t16.773891107623296\t16.621083329448393\t33.285419030007304\n",
      "100\t33.04947346065588\t16.768452186705204\t16.67935748214221\t33.5027168704967\n",
      ">>END_MODULE\n",
      ">>Per sequence GC content\tpass\n",
      "#GC Content\tCount\n",
      "0\t0.0\n",
      "1\t0.0\n",
      "2\t0.0\n",
      "3\t0.0\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      "10\t0.0\n",
      "11\t0.0\n",
      "12\t0.0\n",
      "13\t0.0\n",
      "14\t0.0\n",
      "15\t0.0\n",
      "16\t0.0\n",
      "17\t0.0\n",
      "18\t0.0\n",
      "19\t0.0\n",
      "20\t0.0\n",
      "21\t0.0\n",
      "22\t0.0\n",
      "23\t0.0\n",
      "24\t0.0\n",
      "25\t0.0\n",
      "26\t0.0\n",
      "27\t0.0\n",
      "28\t0.0\n",
      "29\t0.5\n",
      "30\t4.0\n",
      "31\t6.5\n",
      "32\t6.0\n",
      "33\t6.0\n",
      "34\t6.0\n",
      "35\t6.5\n",
      "36\t8.5\n",
      "37\t9.0\n",
      "38\t16.5\n",
      "39\t32.0\n",
      "40\t41.5\n",
      "41\t55.0\n",
      "42\t78.0\n",
      "43\t104.0\n",
      "44\t131.5\n",
      "45\t174.5\n",
      "46\t219.0\n",
      "47\t251.5\n",
      "48\t297.0\n",
      "49\t366.5\n",
      "50\t455.5\n",
      "51\t573.0\n",
      "52\t713.0\n",
      "53\t865.5\n",
      "54\t1093.5\n",
      "55\t1464.0\n",
      "56\t1944.5\n",
      "57\t2577.5\n",
      "58\t3373.5\n",
      "59\t4334.5\n",
      "60\t5514.0\n",
      "61\t6757.5\n",
      "62\t8163.0\n",
      "63\t9629.0\n",
      "64\t10961.5\n",
      "65\t12046.0\n",
      "66\t12786.0\n",
      "67\t13309.0\n",
      "68\t13394.0\n",
      "69\t12948.0\n",
      "70\t12086.5\n",
      "71\t11059.0\n",
      "72\t9890.5\n",
      "73\t8478.0\n",
      "74\t7089.5\n",
      "75\t5679.5\n",
      "76\t4282.5\n",
      "77\t3213.5\n",
      "78\t2329.5\n",
      "79\t1604.0\n",
      "80\t1091.5\n",
      "81\t671.0\n",
      "82\t394.5\n",
      "83\t233.5\n",
      "84\t127.0\n",
      "85\t58.0\n",
      "86\t23.5\n",
      "87\t11.5\n",
      "88\t3.0\n",
      "89\t1.5\n",
      "90\t2.0\n",
      "91\t1.0\n",
      "92\t0.0\n",
      "93\t0.0\n",
      "94\t0.0\n",
      "95\t0.0\n",
      "96\t0.0\n",
      "97\t0.0\n",
      "98\t0.0\n",
      "99\t0.0\n",
      "100\t0.0\n",
      ">>END_MODULE\n",
      ">>Per base N content\tpass\n",
      "#Base\tN-Count\n",
      "1\t0.0\n",
      "2\t0.0\n",
      "3\t0.0\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      "10-11\t0.0\n",
      "12-13\t0.0\n",
      "14-15\t0.0\n",
      "16-17\t0.0\n",
      "18-19\t0.0\n",
      "20-21\t0.0\n",
      "22-23\t0.0\n",
      "24-25\t0.0\n",
      "26-27\t0.0\n",
      "28-29\t0.0\n",
      "30-31\t0.0\n",
      "32-33\t0.0\n",
      "34-35\t0.0\n",
      "36-37\t0.0\n",
      "38-39\t0.0\n",
      "40-41\t0.0\n",
      "42-43\t0.0\n",
      "44-45\t0.0\n",
      "46-47\t0.0\n",
      "48-49\t0.0\n",
      "50-51\t0.0\n",
      "52-53\t0.0\n",
      "54-55\t0.0\n",
      "56-57\t0.0\n",
      "58-59\t0.0\n",
      "60-61\t0.0\n",
      "62-63\t0.0\n",
      "64-65\t0.0\n",
      "66-67\t0.0\n",
      "68-69\t0.0\n",
      "70-71\t0.0\n",
      "72-73\t0.0\n",
      "74-75\t0.0\n",
      "76-77\t0.0\n",
      "78-79\t0.0\n",
      "80-81\t0.0\n",
      "82-83\t0.0\n",
      "84-85\t0.0\n",
      "86-87\t0.0\n",
      "88-89\t0.0\n",
      "90-91\t0.0\n",
      "92-93\t0.0\n",
      "94-95\t0.0\n",
      "96-97\t0.0\n",
      "98-99\t0.0\n",
      "100\t0.0\n",
      ">>END_MODULE\n",
      ">>Sequence Length Distribution\tpass\n",
      "#Length\tCount\n",
      "100\t193053.0\n",
      ">>END_MODULE\n",
      ">>Sequence Duplication Levels\tpass\n",
      "#Total Deduplicated Percentage\t98.5243294134104\n",
      "#Duplication Level\tPercentage of total\n",
      "1\t97.05966508910315\n",
      "2\t2.9073161240498107\n",
      "3\t0.033018786847037844\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      ">10\t0.0\n",
      ">50\t0.0\n",
      ">100\t0.0\n",
      ">500\t0.0\n",
      ">1k\t0.0\n",
      ">5k\t0.0\n",
      ">10k+\t0.0\n",
      ">>END_MODULE\n",
      ">>Overrepresented sequences\tpass\n",
      ">>END_MODULE\n",
      ">>Adapter Content\tpass\n",
      "#Position\tIllumina Universal Adapter\tIllumina Small RNA 3' Adapter\tIllumina Small RNA 5' Adapter\tNextera Transposase Sequence\tPolyA\tPolyG\n",
      "1\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "2\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "3\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "4\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "5\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "6\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "7\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "8\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "9\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "10-11\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "12-13\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "14-15\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "16-17\t0.0\t5.179924683895096E-4\t0.0\t0.0\t0.0\t0.0\n",
      "18-19\t0.0\t0.0010359849367790191\t0.0\t0.0\t0.0\t0.0\n",
      "20-21\t0.0\t0.0010359849367790191\t0.0\t0.0\t0.0\t0.0\n",
      "22-23\t0.0\t0.0010359849367790191\t0.0\t0.0\t0.0\t0.0\n",
      "24-25\t0.0\t0.001294981170973774\t0.0\t0.0\t0.0\t0.0\n",
      "26-27\t0.0\t0.001553977405168529\t0.0\t0.0\t0.0\t0.0\n",
      "28-29\t0.0\t0.001553977405168529\t0.0\t0.0\t0.0\t0.0\n",
      "30-31\t0.0\t0.001553977405168529\t0.0\t0.0\t0.0\t0.0\n",
      "32-33\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "34-35\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "36-37\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "38-39\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "40-41\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "42-43\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "44-45\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "46-47\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "48-49\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "50-51\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "52-53\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "54-55\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "56-57\t0.0\t0.0033669510445318127\t0.0\t0.0\t0.0\t0.0\n",
      "58-59\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "60-61\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "62-63\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "64-65\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "66-67\t0.0\t0.003884943512921322\t0.0\t0.0\t0.0\t0.0\n",
      "68-69\t0.0\t0.004661932215505586\t0.0\t0.0\t0.0\t0.0\n",
      "70-71\t0.0\t0.005179924683895096\t0.0\t0.0\t0.0\t0.0\n",
      ">>END_MODULE\n",
      "\n",
      "file 2: rhodo.art.q20.PE.reads_211157_2_1.fwd_fastqc.zip:\n",
      "##FastQC\t0.12.1\n",
      ">>Basic Statistics\tpass\n",
      "#Measure\tValue\n",
      "Filename\trhodo.art.q20.PE.reads_211157_2_1.fwd.fastq\n",
      "File type\tConventional base calls\n",
      "Encoding\tSanger / Illumina 1.9\n",
      "Total Sequences\t193053\n",
      "Total Bases\t19.3 Mbp\n",
      "Sequences flagged as poor quality\t0\n",
      "Sequence length\t100\n",
      "%GC\t66\n",
      ">>END_MODULE\n",
      ">>Per base sequence quality\tpass\n",
      "#Base\tMean\tMedian\tLower Quartile\tUpper Quartile\t10th Percentile\t90th Percentile\n",
      "1\t50.42625600223773\t51.0\t51.0\t54.0\t47.0\t54.0\n",
      "2\t50.52894800909595\t51.0\t51.0\t54.0\t46.0\t54.0\n",
      "3\t50.32879053938556\t51.0\t51.0\t54.0\t46.0\t54.0\n",
      "4\t53.89255800220665\t57.0\t55.0\t57.0\t48.0\t57.0\n",
      "5\t53.676254707256554\t57.0\t55.0\t57.0\t48.0\t57.0\n",
      "6\t53.67517210299762\t57.0\t55.0\t57.0\t48.0\t57.0\n",
      "7\t53.71687049670298\t57.0\t55.0\t57.0\t49.0\t57.0\n",
      "8\t53.632339305786495\t57.0\t55.0\t57.0\t48.0\t57.0\n",
      "9\t55.05453942699673\t59.0\t55.0\t59.0\t48.0\t59.0\n",
      "10-11\t54.97349432539251\t59.0\t55.0\t59.0\t48.0\t59.0\n",
      "12-13\t54.85529880395539\t59.0\t55.0\t59.0\t47.0\t59.0\n",
      "14-15\t55.89482680921819\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "16-17\t55.727287325242294\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "18-19\t55.70098107773513\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "20-21\t55.642559297187816\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "22-23\t55.59338109223892\t60.0\t56.0\t61.0\t47.0\t61.0\n",
      "24-25\t55.55280156226529\t59.5\t56.0\t61.0\t46.0\t61.0\n",
      "26-27\t55.33956219276572\t59.0\t55.5\t61.0\t45.0\t61.0\n",
      "28-29\t55.19439221353721\t59.0\t55.0\t61.0\t45.0\t61.0\n",
      "30-31\t55.047989412233946\t59.0\t55.0\t61.0\t44.0\t61.0\n",
      "32-33\t54.86606527741087\t59.0\t54.5\t61.0\t44.0\t61.0\n",
      "34-35\t54.67198903927937\t59.0\t54.0\t60.5\t43.5\t61.0\n",
      "36-37\t54.489979435699\t59.0\t54.0\t60.0\t42.0\t61.0\n",
      "38-39\t54.3263663346335\t58.0\t54.0\t60.0\t41.0\t61.0\n",
      "40-41\t54.17223767566419\t58.0\t53.5\t60.0\t39.5\t61.0\n",
      "42-43\t53.97476340694006\t58.0\t53.0\t60.0\t38.5\t61.0\n",
      "44-45\t53.80058325951941\t58.0\t53.0\t60.0\t37.5\t61.0\n",
      "46-47\t53.67100485358942\t58.0\t53.0\t60.0\t36.5\t61.0\n",
      "48-49\t53.44004496174625\t58.0\t53.0\t60.0\t34.5\t61.0\n",
      "50-51\t53.32651655244933\t58.0\t53.0\t60.0\t34.0\t61.0\n",
      "52-53\t53.12265284662761\t58.0\t52.5\t60.0\t32.5\t61.0\n",
      "54-55\t52.65927232418041\t57.0\t52.0\t60.0\t28.0\t61.0\n",
      "56-57\t52.32782707339435\t57.0\t51.0\t60.0\t27.5\t61.0\n",
      "58-59\t52.01091410130897\t56.5\t51.0\t60.0\t26.5\t61.0\n",
      "60-61\t51.797493434445464\t56.0\t51.0\t59.0\t24.0\t61.0\n",
      "62-63\t51.44003201193455\t56.0\t50.5\t59.0\t22.0\t60.5\n",
      "64-65\t51.041892640881\t55.0\t50.0\t59.0\t22.0\t60.0\n",
      "66-67\t50.49370380154672\t55.0\t49.0\t58.5\t22.0\t60.0\n",
      "68-69\t49.959581047691564\t54.5\t48.5\t58.0\t22.0\t60.0\n",
      "70-71\t49.483424241011534\t54.0\t48.0\t57.0\t22.0\t59.5\n",
      "72-73\t48.99674959726086\t54.0\t47.5\t56.5\t22.0\t59.0\n",
      "74-75\t48.62221514299182\t54.0\t46.5\t56.0\t22.0\t59.0\n",
      "76-77\t47.58507508300829\t52.5\t46.0\t55.0\t22.0\t57.0\n",
      "78-79\t48.239677705086166\t54.0\t47.0\t55.0\t22.0\t57.0\n",
      "80-81\t48.06770938550554\t54.0\t47.0\t55.0\t22.0\t57.0\n",
      "82-83\t47.82851341341497\t54.0\t46.5\t55.0\t22.0\t56.5\n",
      "84-85\t47.44304672810057\t53.0\t46.0\t55.0\t22.0\t56.0\n",
      "86-87\t47.02018616649314\t53.0\t45.5\t55.0\t22.0\t56.0\n",
      "88-89\t46.609412959135575\t53.0\t45.0\t55.0\t22.0\t55.0\n",
      "90-91\t46.146970003056154\t52.5\t44.5\t55.0\t22.0\t55.0\n",
      "92-93\t45.50171455507037\t52.0\t41.5\t54.5\t22.0\t55.0\n",
      "94-95\t44.92707443033778\t52.0\t38.5\t54.0\t22.0\t55.0\n",
      "96-97\t44.322284553982584\t52.0\t28.5\t54.0\t22.0\t55.0\n",
      "98-99\t43.466910641119284\t51.0\t22.0\t54.0\t22.0\t55.0\n",
      "100\t45.04022729509513\t51.0\t31.0\t54.0\t31.0\t55.0\n",
      ">>END_MODULE\n",
      ">>Per sequence quality scores\tpass\n",
      "#Quality\tCount\n",
      "45\t3.0\n",
      "46\t43.0\n",
      "47\t287.0\n",
      "48\t2362.0\n",
      "49\t12266.0\n",
      "50\t39684.0\n",
      "51\t67212.0\n",
      "52\t53493.0\n",
      "53\t16149.0\n",
      "54\t1531.0\n",
      "55\t23.0\n",
      ">>END_MODULE\n",
      ">>Per base sequence content\tpass\n",
      "#Base\tG\tA\tT\tC\n",
      "1\t33.40481629397109\t16.693343278788728\t16.566435124033298\t33.33540530320689\n",
      "2\t33.2312888170606\t16.832165260317115\t16.65967376834341\t33.27687215427888\n",
      "3\t33.34213920529596\t16.65656581353307\t16.731156728981162\t33.27013825218981\n",
      "4\t33.26703029737948\t16.737890631070222\t16.652421873785954\t33.342657197764346\n",
      "5\t33.51307671986449\t16.648277934038838\t16.637400092202657\t33.20124525389401\n",
      "6\t33.19502934427334\t16.812999538986702\t16.709401045308798\t33.28257007143116\n",
      "7\t33.35405303206891\t16.74462453315929\t16.65552982859629\t33.245792606175506\n",
      "8\t33.27842613168405\t16.77259612645232\t16.823359388354493\t33.12561835350914\n",
      "9\t33.19451135180495\t16.77415010385749\t16.725976804297265\t33.305361740040304\n",
      "10-11\t33.26521732374012\t16.67832149720543\t16.689976327744194\t33.36648485131026\n",
      "12-13\t33.26185037269558\t16.807560618068614\t16.774927092560073\t33.15566191667573\n",
      "14-15\t33.418284098149215\t16.710437030245583\t16.651903881317565\t33.21937499028764\n",
      "16-17\t33.36104593039217\t16.557370255836478\t16.759387318508388\t33.32219649526296\n",
      "18-19\t33.3149446007055\t16.734782676259886\t16.619788348277417\t33.33048437475719\n",
      "20-21\t33.32737641994685\t16.784509953225278\t16.621342325682583\t33.26677130114528\n",
      "22-23\t33.35819697181603\t16.585341849129513\t16.838122173703592\t33.21833900535086\n",
      "24-25\t33.252526508264566\t16.723645838189512\t16.674695549926703\t33.34913210361921\n",
      "26-27\t33.40248532786333\t16.735300668728275\t16.53121163618281\t33.33100236722558\n",
      "28-29\t33.24786457604906\t16.66588967796408\t16.673659564989926\t33.412586180996925\n",
      "30-31\t33.28593702247569\t16.695415248662286\t16.614349427359326\t33.40429830150269\n",
      "32-33\t33.32245549149715\t16.693343278788728\t16.5858598415979\t33.398341388116215\n",
      "34-35\t33.322714487731346\t16.546751410234496\t16.7886538929724\t33.341880209061756\n",
      "36-37\t33.14323009743438\t16.65656581353307\t16.743847544456703\t33.45635654457585\n",
      "38-39\t33.317016570579064\t16.654234847425318\t16.71354498505592\t33.3152035969397\n",
      "40-41\t33.360009945455396\t16.704221120624908\t16.6651126892615\t33.270656244658205\n",
      "42-43\t33.27195122582918\t16.68764536163644\t16.619270355809025\t33.42113305672535\n",
      "44-45\t33.34680113751146\t16.659414772109212\t16.72519981559468\t33.268584274784644\n",
      "46-47\t33.28645501494408\t16.68971733151\t16.69127130891517\t33.332556344630746\n",
      "48-49\t33.30613872874289\t16.737372638601837\t16.64775994157045\t33.30872869108483\n",
      "50-51\t33.338772254251424\t16.831647267848727\t16.70007718087779\t33.12950329702206\n",
      "52-53\t33.21859800158506\t16.703703128156516\t16.73607765743086\t33.34162121282757\n",
      "54-55\t33.19995027272304\t16.75731534863483\t16.68971733151\t33.353017047132134\n",
      "56-57\t33.26469933127172\t16.694379263725505\t16.810927569113147\t33.22999383588963\n",
      "58-59\t33.33618229190948\t16.668997632774417\t16.709401045308798\t33.285419030007304\n",
      "60-61\t33.16058284512543\t16.75679735616644\t16.72778977793663\t33.354830020771495\n",
      "62-63\t33.1768996078797\t16.740480593412173\t16.737372638601837\t33.34524716010629\n",
      "64-65\t33.26469933127172\t16.728307770405017\t16.743847544456703\t33.263145353866555\n",
      "66-67\t33.27687215427888\t16.653457858722735\t16.69075331644678\t33.37891667055161\n",
      "68-69\t33.24786457604906\t16.72519981559468\t16.731415725215356\t33.2955198831409\n",
      "70-71\t33.354830020771495\t16.672105587584756\t16.72571780806307\t33.24734658358067\n",
      "72-73\t33.32297348396555\t16.63144317881618\t16.81481251262607\t33.23077082459221\n",
      "74-75\t33.3253044500733\t16.741257582114756\t16.660709753280187\t33.27272821453176\n",
      "76-77\t33.30691571744547\t16.77958902477558\t16.65708380600146\t33.25641145177749\n",
      "78-79\t33.304066758869325\t16.58896779640824\t16.82646734316483\t33.280498101557605\n",
      "80-81\t33.31364961953453\t16.735300668728275\t16.616421397232884\t33.33462831450431\n",
      "82-83\t33.29888683418543\t16.718983905974007\t16.632479163752958\t33.3496500960876\n",
      "84-85\t33.36700284377866\t16.76223627708453\t16.64439299052592\t33.226367888610895\n",
      "86-87\t33.240094689023216\t16.696710229833258\t16.732710706386328\t33.33048437475719\n",
      "88-89\t33.33721827684625\t16.71639394363206\t16.74462453315929\t33.2017632463624\n",
      "90-91\t33.19218038569719\t16.752135423950936\t16.692825286320335\t33.36285890403153\n",
      "92-93\t33.41673012074404\t16.723127845721123\t16.554780293494535\t33.305361740040304\n",
      "94-95\t33.26262736139817\t16.808078610537002\t16.617716378403856\t33.311577649660975\n",
      "96-97\t33.268066282316255\t16.66200473445116\t16.764826239426476\t33.3051027438061\n",
      "98-99\t33.15307195433378\t16.767934194236815\t16.752135423950936\t33.32685842747846\n",
      "100\t33.24268465136517\t16.666407670432474\t16.693343278788728\t33.39756439941363\n",
      ">>END_MODULE\n",
      ">>Per sequence GC content\tpass\n",
      "#GC Content\tCount\n",
      "0\t0.0\n",
      "1\t0.0\n",
      "2\t0.0\n",
      "3\t0.0\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      "10\t0.0\n",
      "11\t0.0\n",
      "12\t0.0\n",
      "13\t0.0\n",
      "14\t0.0\n",
      "15\t0.0\n",
      "16\t0.0\n",
      "17\t0.0\n",
      "18\t0.0\n",
      "19\t0.0\n",
      "20\t0.0\n",
      "21\t0.0\n",
      "22\t0.0\n",
      "23\t0.0\n",
      "24\t0.0\n",
      "25\t0.0\n",
      "26\t0.0\n",
      "27\t0.0\n",
      "28\t0.0\n",
      "29\t0.0\n",
      "30\t2.0\n",
      "31\t7.0\n",
      "32\t10.5\n",
      "33\t8.5\n",
      "34\t7.5\n",
      "35\t8.0\n",
      "36\t7.0\n",
      "37\t11.5\n",
      "38\t20.0\n",
      "39\t25.0\n",
      "40\t44.5\n",
      "41\t67.0\n",
      "42\t80.0\n",
      "43\t91.5\n",
      "44\t118.0\n",
      "45\t168.0\n",
      "46\t214.5\n",
      "47\t268.0\n",
      "48\t321.0\n",
      "49\t377.0\n",
      "50\t460.0\n",
      "51\t559.0\n",
      "52\t676.5\n",
      "53\t848.5\n",
      "54\t1064.0\n",
      "55\t1423.0\n",
      "56\t1910.5\n",
      "57\t2505.5\n",
      "58\t3366.0\n",
      "59\t4353.0\n",
      "60\t5474.5\n",
      "61\t6780.0\n",
      "62\t8181.5\n",
      "63\t9598.0\n",
      "64\t10884.5\n",
      "65\t12022.5\n",
      "66\t12985.5\n",
      "67\t13463.0\n",
      "68\t13348.5\n",
      "69\t12959.5\n",
      "70\t12207.5\n",
      "71\t11147.5\n",
      "72\t9880.0\n",
      "73\t8448.5\n",
      "74\t7001.0\n",
      "75\t5576.5\n",
      "76\t4333.0\n",
      "77\t3236.5\n",
      "78\t2315.5\n",
      "79\t1602.0\n",
      "80\t1050.5\n",
      "81\t663.5\n",
      "82\t398.5\n",
      "83\t230.5\n",
      "84\t126.0\n",
      "85\t66.0\n",
      "86\t34.0\n",
      "87\t11.0\n",
      "88\t3.5\n",
      "89\t0.5\n",
      "90\t0.0\n",
      "91\t0.5\n",
      "92\t0.5\n",
      "93\t0.0\n",
      "94\t0.0\n",
      "95\t0.0\n",
      "96\t0.0\n",
      "97\t0.0\n",
      "98\t0.0\n",
      "99\t0.0\n",
      "100\t0.0\n",
      ">>END_MODULE\n",
      ">>Per base N content\tpass\n",
      "#Base\tN-Count\n",
      "1\t0.0\n",
      "2\t0.0\n",
      "3\t0.0\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      "10-11\t0.0\n",
      "12-13\t0.0\n",
      "14-15\t0.0\n",
      "16-17\t0.0\n",
      "18-19\t0.0\n",
      "20-21\t0.0\n",
      "22-23\t0.0\n",
      "24-25\t0.0\n",
      "26-27\t0.0\n",
      "28-29\t0.0\n",
      "30-31\t0.0\n",
      "32-33\t0.0\n",
      "34-35\t0.0\n",
      "36-37\t0.0\n",
      "38-39\t0.0\n",
      "40-41\t0.0\n",
      "42-43\t0.0\n",
      "44-45\t0.0\n",
      "46-47\t0.0\n",
      "48-49\t0.0\n",
      "50-51\t0.0\n",
      "52-53\t0.0\n",
      "54-55\t0.0\n",
      "56-57\t0.0\n",
      "58-59\t0.0\n",
      "60-61\t0.0\n",
      "62-63\t0.0\n",
      "64-65\t0.0\n",
      "66-67\t0.0\n",
      "68-69\t0.0\n",
      "70-71\t0.0\n",
      "72-73\t0.0\n",
      "74-75\t0.0\n",
      "76-77\t0.0\n",
      "78-79\t0.0\n",
      "80-81\t0.0\n",
      "82-83\t0.0\n",
      "84-85\t0.0\n",
      "86-87\t0.0\n",
      "88-89\t0.0\n",
      "90-91\t0.0\n",
      "92-93\t0.0\n",
      "94-95\t0.0\n",
      "96-97\t0.0\n",
      "98-99\t0.0\n",
      "100\t0.0\n",
      ">>END_MODULE\n",
      ">>Sequence Length Distribution\tpass\n",
      "#Length\tCount\n",
      "100\t193053.0\n",
      ">>END_MODULE\n",
      ">>Sequence Duplication Levels\tpass\n",
      "#Total Deduplicated Percentage\t98.51099750971424\n",
      "#Duplication Level\tPercentage of total\n",
      "1\t97.02969767266319\n",
      "2\t2.947194367632699\n",
      "3\t0.023107959704111326\n",
      "4\t0.0\n",
      "5\t0.0\n",
      "6\t0.0\n",
      "7\t0.0\n",
      "8\t0.0\n",
      "9\t0.0\n",
      ">10\t0.0\n",
      ">50\t0.0\n",
      ">100\t0.0\n",
      ">500\t0.0\n",
      ">1k\t0.0\n",
      ">5k\t0.0\n",
      ">10k+\t0.0\n",
      ">>END_MODULE\n",
      ">>Overrepresented sequences\tpass\n",
      ">>END_MODULE\n",
      ">>Adapter Content\tpass\n",
      "#Position\tIllumina Universal Adapter\tIllumina Small RNA 3' Adapter\tIllumina Small RNA 5' Adapter\tNextera Transposase Sequence\tPolyA\tPolyG\n",
      "1\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "2\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "3\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "4\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "5\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "6\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "7\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "8\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "9\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "10-11\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "12-13\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "14-15\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "16-17\t0.0\t2.589962341947548E-4\t0.0\t0.0\t0.0\t0.0\n",
      "18-19\t0.0\t7.769887025842644E-4\t0.0\t0.0\t0.0\t0.0\n",
      "20-21\t0.0\t0.001553977405168529\t0.0\t0.0\t0.0\t0.0\n",
      "22-23\t0.0\t0.0018129736393632835\t0.0\t0.0\t0.0\t0.0\n",
      "24-25\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "26-27\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "28-29\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "30-31\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "32-33\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "34-35\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "36-37\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "38-39\t0.0\t0.0020719698735580383\t0.0\t0.0\t0.0\t0.0\n",
      "40-41\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "42-43\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "44-45\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "46-47\t0.0\t0.002589962341947548\t0.0\t0.0\t0.0\t0.0\n",
      "48-49\t0.0\t0.0028489585761423027\t0.0\t0.0\t0.0\t0.0\n",
      "50-51\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "52-53\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "54-55\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "56-57\t0.0\t0.003107954810337058\t0.0\t0.0\t0.0\t0.0\n",
      "58-59\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "60-61\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "62-63\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "64-65\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "66-67\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "68-69\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "70-71\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "72-73\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "74-75\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "76-77\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "78-79\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "80-81\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "82-83\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "84-85\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "86-87\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      "88-89\t0.0\t0.0036259472787265675\t0.0\t0.0\t0.0\t0.0\n",
      ">>END_MODULE\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mAnalyze the given report and derive some biological insight into the result.\n",
      "            If the report is not in JSON format, then interpret the document as-is.\n",
      "            If it is in JSON format, The report may contain content in 3 categories.\n",
      "            1. \"message\": this is a brief message describing the outcome of the report\n",
      "            2. \"direct html\": this is HTML-formatted information meant to be displayed to the user, and might be a brief summary of the full report.\n",
      "            3. \"html report\": this is one or more full HTML-formatted documents containing the report information.\n",
      "\n",
      "            Regardless of the format, it may be plain text, which can be interpreted as-is. It may also\n",
      "            also be formatted as HTML. If so, read the HTML document, including any base-64 encoded images for interpretation.\n",
      "\n",
      "            After interpretation, analyze the report, and summarize the findings with a biological interpretation. Write a brief summary, including\n",
      "            bullet points when appropriate, formatted in markdown. Your final answer MUST be the summary of the report.\n",
      "\n",
      "            If there was no report, the report is null, or an empty string, return a note saying that there is no report to analyze.\n",
      "            If the previous task ended with an error, or another note saying that there is no report to analyze,\n",
      "            just return a note saying so. Otherwise, return the summary of the report.\n",
      "            \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "### Summary of FastQC Report\n",
      "\n",
      "#### Basic Statistics\n",
      "- **Filename**: rhodo.art.q20.PE.reads_211157_2_1.rev.fastq and rhodo.art.q20.PE.reads_211157_2_1.fwd.fastq\n",
      "- **File Type**: Conventional base calls\n",
      "- **Encoding**: Sanger / Illumina 1.9\n",
      "- **Total Sequences**: 193,053 for each file\n",
      "- **Total Bases**: 19.3 Mbp for each file\n",
      "- **Sequences Flagged as Poor Quality**: 0\n",
      "- **Sequence Length**: 100\n",
      "- **%GC**: 66\n",
      "\n",
      "#### Per Base Sequence Quality\n",
      "- The quality scores are high across all bases, with mean scores above 40, indicating good quality reads.\n",
      "\n",
      "#### Per Sequence Quality Scores\n",
      "- The majority of sequences have high-quality scores, with most scores above 45.\n",
      "\n",
      "#### Per Base Sequence Content\n",
      "- The base composition is consistent across the length of the reads, with a balanced representation of G, A, T, and C bases.\n",
      "\n",
      "#### Per Sequence GC Content\n",
      "- The GC content is consistent with the expected 66%, with a normal distribution around this value.\n",
      "\n",
      "#### Per Base N Content\n",
      "- No N bases were detected, indicating no ambiguous base calls.\n",
      "\n",
      "#### Sequence Length Distribution\n",
      "- All sequences are of uniform length (100 bases), as expected.\n",
      "\n",
      "#### Sequence Duplication Levels\n",
      "- The deduplicated percentage is high (98.5%), indicating low levels of sequence duplication.\n",
      "\n",
      "#### Overrepresented Sequences\n",
      "- No overrepresented sequences were detected, suggesting no contamination or bias.\n",
      "\n",
      "#### Adapter Content\n",
      "- No adapter contamination was detected, indicating clean sequencing data.\n",
      "\n",
      "### Biological Interpretation\n",
      "- The sequencing data is of high quality, with consistent base composition and no significant issues detected. The high GC content is typical for the organism being studied. The data is suitable for downstream analysis, such as genome assembly or variant calling, without the need for additional quality control measures.\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mSave the analysis by adding a markdown cell to the Narrative with id 211157. The markdown text must\n",
      "            be the analysis text. If not successful, say so and stop. If an output object was created, ensure that it has both an UPA and name.\n",
      "            In the end, return the results of the job completion task with both UPA and name for output object. The return result must be normalized to contain\n",
      "            both the UPA and output object name, if either are present. If the app has neither an output object name or UPA, both of these\n",
      "            fields may be None.\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92madd-markdown-cell\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"markdown_text\\\": \\\"### Summary of FastQC Report\\\\n\\\\n#### Basic Statistics\\\\n- **Filename**: rhodo.art.q20.PE.reads_211157_2_1.rev.fastq and rhodo.art.q20.PE.reads_211157_2_1.fwd.fastq\\\\n- **File Type**: Conventional base calls\\\\n- **Encoding**: Sanger / Illumina 1.9\\\\n- **Total Sequences**: 193,053 for each file\\\\n- **Total Bases**: 19.3 Mbp for each file\\\\n- **Sequences Flagged as Poor Quality**: 0\\\\n- **Sequence Length**: 100\\\\n- **%GC**: 66\\\\n\\\\n#### Per Base Sequence Quality\\\\n- The quality scores are high across all bases, with mean scores above 40, indicating good quality reads.\\\\n\\\\n#### Per Sequence Quality Scores\\\\n- The majority of sequences have high-quality scores, with most scores above 45.\\\\n\\\\n#### Per Base Sequence Content\\\\n- The base composition is consistent across the length of the reads, with a balanced representation of G, A, T, and C bases.\\\\n\\\\n#### Per Sequence GC Content\\\\n- The GC content is consistent with the expected 66%, with a normal distribution around this value.\\\\n\\\\n#### Per Base N Content\\\\n- No N bases were detected, indicating no ambiguous base calls.\\\\n\\\\n#### Sequence Length Distribution\\\\n- All sequences are of uniform length (100 bases), as expected.\\\\n\\\\n#### Sequence Duplication Levels\\\\n- The deduplicated percentage is high (98.5%), indicating low levels of sequence duplication.\\\\n\\\\n#### Overrepresented Sequences\\\\n- No overrepresented sequences were detected, suggesting no contamination or bias.\\\\n\\\\n#### Adapter Content\\\\n- No adapter contamination was detected, indicating clean sequencing data.\\\\n\\\\n### Biological Interpretation\\\\n- The sequencing data is of high quality, with consistent base composition and no significant issues detected. The high GC content is typical for the organism being studied. The data is suitable for downstream analysis, such as genome assembly or variant calling, without the need for additional quality control measures.\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "success\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"job_id\": \"67e2f276138d24fd8da1908b\",\n",
      "  \"job_status\": \"completed\",\n",
      "  \"created_objects\": [],\n",
      "  \"job_error\": null,\n",
      "  \"report_upa\": \"211157/3/1\"\n",
      "}\u001b[00m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mYou are a workflow runner, your role is to efficiently run KBase workflows.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mRun KBase app\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"app_id\\\": \\\"kb_fastqc/runFastQC\\\", \\\"input_object_upa\\\": \\\"211157/2/1\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "job_id='67e2f276138d24fd8da1908b' job_status='completed' created_objects=[] job_error=None report_upa='211157/3/1'\u001b[00m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overriding of current TracerProvider is not allowed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            From the given KBase app id, kb_trimmomatic/run_trimmomatic, fetch the list of parameters needed to run it. Use the App and Job manager agent\n",
      "            for assistance. Using the data object with UPA \"211157/2/1\", populate a dictionary\n",
      "            with the parameters where the keys are parameter ids, and values are the proper parameter values, or their\n",
      "            default values if no value can be found or calculated.\n",
      "            Any input object parameter must be the input object UPA.\n",
      "            Be sure to make sure there is a non-null value for any parameter that is not optional.\n",
      "            Any parameter that has a true value for \"is_output_object\" must have a valid name for the new object. The new object name should be based on\n",
      "            the input object name, not its UPA. If the input object name is not available, the Workspace Manager can assist.\n",
      "            If the parameter type is 'dropdown', use the allowed 'name' option to determine what should be used, but only set the\n",
      "            'value' associated with that name, or the default value if any.\n",
      "            Only alphanumeric characters and underscores are allowed in new object names.\n",
      "            Return the dictionary of inputs, the app id, and the\n",
      "            narrative id 211157 for use in the next task. Do not add comments or other text. If the parameters are rejected, examine the reason why and\n",
      "            reform them. The dictionary of inputs and the app id must not be combined into a single dictionary.\n",
      "            \u001b[00m\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mWhat are the parameters needed to run the KBase app with the id 'kb_trimmomatic/run_trimmomatic'?\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mget-app-parameters\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"app_id\\\": \\\"kb_trimmomatic/run_trimmomatic\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "{\"input_reads_ref\": {\"id\": \"input_reads_ref\", \"ui_name\": \"Read library (or Set) object <font color=red>*</font>\", \"short_hint\": \"The read library or read library set for trimming.\", \"is_output_object\": false, \"type\": \"data_object\", \"allowed_data_type\": [\"KBaseRNASeq.RNASeqSampleSet\", \"KBaseSets.ReadsSet\", \"KBaseFile.PairedEndLibrary\", \"KBaseFile.SingleEndLibrary\", \"KBaseAssembly.PairedEndLibrary\", \"KBaseAssembly.SingleEndLibrary\"], \"multiple\": false, \"optional\": false, \"default_value\": null}, \"output_reads_name\": {\"id\": \"output_reads_name\", \"ui_name\": \"Output library (or Set) name\", \"short_hint\": \"Base name for the output read library or read library set object(s).\", \"is_output_object\": true, \"type\": \"data_object\", \"allowed_data_type\": [\"KBaseSets.ReadsSet\", \"KBaseFile.PairedEndLibrary\", \"KBaseFile.SingleEndLibrary\"], \"multiple\": false, \"optional\": false, \"default_value\": null}, \"translate_to_phred33\": {\"id\": \"translate_to_phred33\", \"ui_name\": \"Translate quality encoding from phred64 to phred33\", \"short_hint\": \"Modernize and standardize phred64 (Illumina <= 1.7) read quality encoding format to phred33 (Sanger, 454, SRA, Illumina >= 1.8). Does nothing if already phred33.\", \"is_output_object\": false, \"type\": \"dropdown\", \"allowed\": [{\"name\": \"no\", \"value\": \"0\"}, {\"name\": \"yes\", \"value\": \"1\"}], \"multiple\": false, \"optional\": true, \"default_value\": \"1\"}, \"crop_length\": {\"id\": \"crop_length\", \"ui_name\": \"Post tail crop length\", \"short_hint\": \"The number of bases to keep, from the start of the read (e.g. tail crop 10 bases from a 150 base read: enter \\\"140\\\"). Removes bases regardless of quality from the end of a read so that the read has the specified length after this step has been performed. Steps performed after CROP might further shorten the read.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": null}, \"head_crop_length\": {\"id\": \"head_crop_length\", \"ui_name\": \"Head crop length\", \"short_hint\": \"The number of bases to remove from the start of a read (e.g. head crop 10 bases: enter \\\"10\\\"). Removes the specified number of bases, regardless of quality, from the beginning of the read.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": 0}, \"leading_min_quality\": {\"id\": \"leading_min_quality\", \"ui_name\": \"Leading minimum quality\", \"short_hint\": \"Specifies the minimum quality required to keep a base at the start of a read. Removes low-quality bases from the beginning. As long as a base has a value below this threshold, the base is removed and the next base is investigated.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": 3}, \"trailing_min_quality\": {\"id\": \"trailing_min_quality\", \"ui_name\": \"Trailing minimum quality\", \"short_hint\": \"Specifies the minimum quality required to keep a base at the end of a read. Remove low-quality bases from the end. As long as a base has a value below this threshold, it is removed and the next base will be investigated (if Trimmomatic starts from the 3 prime end, it is the base preceding the removed one). This approach is used for removing the special Illumina low-quality segment regions (which are marked with a quality score of 2), but we recommend Sliding window or MaxInfro instead.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": 3}, \"min_length\": {\"id\": \"min_length\", \"ui_name\": \"Minimum read length\", \"short_hint\": \"Specifies the minimum length of reads to be kept. This module removes reads that fall below the specified minimum length. Reads removed by this step will be counted and included in the \\\"dropped reads\\\" count presented in the Trimmomatic summary.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": 36}, \"adapter_clip\": {\"id\": \"adapter_clip\", \"ui_name\": \"Adapter clipping options\", \"short_hint\": \"Configure if you wish to remove standard adapters.\", \"params\": {\"adapterFa\": {\"id\": \"adapterFa\", \"ui_name\": \"Adapters\", \"short_hint\": \"The reference adapter sequences for trimming.\", \"is_output_object\": false, \"type\": \"dropdown\", \"allowed\": [{\"name\": \"TruSeq3-PE\", \"value\": \"TruSeq3-PE.fa\"}, {\"name\": \"TruSeq2-PE\", \"value\": \"TruSeq2-PE.fa\"}, {\"name\": \"TruSeq3-PE-2\", \"value\": \"TruSeq3-PE-2.fa\"}, {\"name\": \"TruSeq3-SE\", \"value\": \"TruSeq3-SE.fa\"}, {\"name\": \"TruSeq2-SE\", \"value\": \"TruSeq2-SE.fa\"}, {\"name\": \"NexteraPE-PE\", \"value\": \"NexteraPE-PE.fa\"}], \"multiple\": false, \"optional\": true, \"default_value\": null}, \"seed_mismatches\": {\"id\": \"seed_mismatches\", \"ui_name\": \"Seed mismatches\", \"short_hint\": \"Specifies the maximum mismatch count, which will still allow a full match to be performed.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": 2}, \"palindrome_clip_threshold\": {\"id\": \"palindrome_clip_threshold\", \"ui_name\": \"Palindrome clip threshold\", \"short_hint\": \"Specifies how accurate the match between the two 'adapter ligated' reads must be for PE palindrome read alignment.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": 30}, \"simple_clip_threshold\": {\"id\": \"simple_clip_threshold\", \"ui_name\": \"Simple clip threshold\", \"short_hint\": \"Specifies how accurate the match between any adapter sequence must be against a read.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": 10}}, \"optional\": true, \"as_list\": false, \"type\": \"group\"}, \"sliding_window\": {\"id\": \"sliding_window\", \"ui_name\": \"Sliding window options\", \"short_hint\": \"Configure quality window options.\", \"params\": {\"sliding_window_size\": {\"id\": \"sliding_window_size\", \"ui_name\": \"Sliding window size\", \"short_hint\": \"Specifies the number of bases to average across. Perform a sliding window trimming, cutting once the average quality within the window falls below a threshold. By considering multiple bases, a single poor quality base will not cause the removal of high-quality data later in the red.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": 4}, \"sliding_window_min_quality\": {\"id\": \"sliding_window_min_quality\", \"ui_name\": \"Sliding window minimum quality\", \"short_hint\": \"Specifies the average quality required. Perform a sliding window trimming, cutting once the average quality within the window falls below a threshold. By considering multipe bases, a single poor quality base will not cause the removal of high-quality data later in the read.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": false, \"optional\": true, \"default_value\": 15}}, \"optional\": false, \"as_list\": false, \"type\": \"group\"}}\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "The parameters needed to run the KBase app 'kb_trimmomatic/run_trimmomatic' are as follows:\n",
      "\n",
      "1. `input_reads_ref`: \n",
      "   - UI Name: Read library (or Set) object\n",
      "   - Type: data_object\n",
      "   - Allowed Data Types: KBaseRNASeq.RNASeqSampleSet, KBaseSets.ReadsSet, KBaseFile.PairedEndLibrary, KBaseFile.SingleEndLibrary, KBaseAssembly.PairedEndLibrary, KBaseAssembly.SingleEndLibrary\n",
      "   - Optional: No\n",
      "\n",
      "2. `output_reads_name`: \n",
      "   - UI Name: Output library (or Set) name\n",
      "   - Type: data_object\n",
      "   - Allowed Data Types: KBaseSets.ReadsSet, KBaseFile.PairedEndLibrary, KBaseFile.SingleEndLibrary\n",
      "   - Optional: No\n",
      "\n",
      "3. `translate_to_phred33`: \n",
      "   - UI Name: Translate quality encoding from phred64 to phred33\n",
      "   - Type: dropdown\n",
      "   - Allowed Values: no (0), yes (1)\n",
      "   - Default Value: 1\n",
      "   - Optional: Yes\n",
      "\n",
      "4. `crop_length`: \n",
      "   - UI Name: Post tail crop length\n",
      "   - Type: int\n",
      "   - Optional: Yes\n",
      "\n",
      "5. `head_crop_length`: \n",
      "   - UI Name: Head crop length\n",
      "   - Type: int\n",
      "   - Default Value: 0\n",
      "   - Optional: Yes\n",
      "\n",
      "6. `leading_min_quality`: \n",
      "   - UI Name: Leading minimum quality\n",
      "   - Type: int\n",
      "   - Default Value: 3\n",
      "   - Optional: Yes\n",
      "\n",
      "7. `trailing_min_quality`: \n",
      "   - UI Name: Trailing minimum quality\n",
      "   - Type: int\n",
      "   - Default Value: 3\n",
      "   - Optional: Yes\n",
      "\n",
      "8. `min_length`: \n",
      "   - UI Name: Minimum read length\n",
      "   - Type: int\n",
      "   - Default Value: 36\n",
      "   - Optional: Yes\n",
      "\n",
      "9. `adapter_clip`: \n",
      "   - UI Name: Adapter clipping options\n",
      "   - Type: group\n",
      "   - Optional: Yes\n",
      "   - Parameters:\n",
      "     - `adapterFa`: \n",
      "       - UI Name: Adapters\n",
      "       - Type: dropdown\n",
      "       - Allowed Values: TruSeq3-PE, TruSeq2-PE, TruSeq3-PE-2, TruSeq3-SE, TruSeq2-SE, NexteraPE-PE\n",
      "     - `seed_mismatches`: \n",
      "       - UI Name: Seed mismatches\n",
      "       - Type: int\n",
      "       - Default Value: 2\n",
      "     - `palindrome_clip_threshold`: \n",
      "       - UI Name: Palindrome clip threshold\n",
      "       - Type: int\n",
      "       - Default Value: 30\n",
      "     - `simple_clip_threshold`: \n",
      "       - UI Name: Simple clip threshold\n",
      "       - Type: int\n",
      "       - Default Value: 10\n",
      "\n",
      "10. `sliding_window`: \n",
      "    - UI Name: Sliding window options\n",
      "    - Type: group\n",
      "    - Optional: No\n",
      "    - Parameters:\n",
      "      - `sliding_window_size`: \n",
      "        - UI Name: Sliding window size\n",
      "        - Type: int\n",
      "        - Default Value: 4\n",
      "      - `sliding_window_min_quality`: \n",
      "        - UI Name: Sliding window minimum quality\n",
      "        - Type: int\n",
      "        - Default Value: 15\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mI need to first fetch the list of parameters needed to run the KBase app with the id \"kb_trimmomatic/run_trimmomatic\" using the Job and App Manager.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mAsk question to coworker\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"question\\\": \\\"What are the parameters needed to run the KBase app with the id 'kb_trimmomatic/run_trimmomatic'?\\\", \\\"context\\\": \\\"I need to fetch the list of parameters required to run the KBase app with the id 'kb_trimmomatic/run_trimmomatic'. This information is crucial for populating a dictionary with the appropriate parameter values for a data object with UPA '211157/2/1'.\\\", \\\"coworker\\\": \\\"Job and App Manager\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "The parameters needed to run the KBase app 'kb_trimmomatic/run_trimmomatic' are as follows:\n",
      "\n",
      "1. `input_reads_ref`: \n",
      "   - UI Name: Read library (or Set) object\n",
      "   - Type: data_object\n",
      "   - Allowed Data Types: KBaseRNASeq.RNASeqSampleSet, KBaseSets.ReadsSet, KBaseFile.PairedEndLibrary, KBaseFile.SingleEndLibrary, KBaseAssembly.PairedEndLibrary, KBaseAssembly.SingleEndLibrary\n",
      "   - Optional: No\n",
      "\n",
      "2. `output_reads_name`: \n",
      "   - UI Name: Output library (or Set) name\n",
      "   - Type: data_object\n",
      "   - Allowed Data Types: KBaseSets.ReadsSet, KBaseFile.PairedEndLibrary, KBaseFile.SingleEndLibrary\n",
      "   - Optional: No\n",
      "\n",
      "3. `translate_to_phred33`: \n",
      "   - UI Name: Translate quality encoding from phred64 to phred33\n",
      "   - Type: dropdown\n",
      "   - Allowed Values: no (0), yes (1)\n",
      "   - Default Value: 1\n",
      "   - Optional: Yes\n",
      "\n",
      "4. `crop_length`: \n",
      "   - UI Name: Post tail crop length\n",
      "   - Type: int\n",
      "   - Optional: Yes\n",
      "\n",
      "5. `head_crop_length`: \n",
      "   - UI Name: Head crop length\n",
      "   - Type: int\n",
      "   - Default Value: 0\n",
      "   - Optional: Yes\n",
      "\n",
      "6. `leading_min_quality`: \n",
      "   - UI Name: Leading minimum quality\n",
      "   - Type: int\n",
      "   - Default Value: 3\n",
      "   - Optional: Yes\n",
      "\n",
      "7. `trailing_min_quality`: \n",
      "   - UI Name: Trailing minimum quality\n",
      "   - Type: int\n",
      "   - Default Value: 3\n",
      "   - Optional: Yes\n",
      "\n",
      "8. `min_length`: \n",
      "   - UI Name: Minimum read length\n",
      "   - Type: int\n",
      "   - Default Value: 36\n",
      "   - Optional: Yes\n",
      "\n",
      "9. `adapter_clip`: \n",
      "   - UI Name: Adapter clipping options\n",
      "   - Type: group\n",
      "   - Optional: Yes\n",
      "   - Parameters:\n",
      "     - `adapterFa`: \n",
      "       - UI Name: Adapters\n",
      "       - Type: dropdown\n",
      "       - Allowed Values: TruSeq3-PE, TruSeq2-PE, TruSeq3-PE-2, TruSeq3-SE, TruSeq2-SE, NexteraPE-PE\n",
      "     - `seed_mismatches`: \n",
      "       - UI Name: Seed mismatches\n",
      "       - Type: int\n",
      "       - Default Value: 2\n",
      "     - `palindrome_clip_threshold`: \n",
      "       - UI Name: Palindrome clip threshold\n",
      "       - Type: int\n",
      "       - Default Value: 30\n",
      "     - `simple_clip_threshold`: \n",
      "       - UI Name: Simple clip threshold\n",
      "       - Type: int\n",
      "       - Default Value: 10\n",
      "\n",
      "10. `sliding_window`: \n",
      "    - UI Name: Sliding window options\n",
      "    - Type: group\n",
      "    - Optional: No\n",
      "    - Parameters:\n",
      "      - `sliding_window_size`: \n",
      "        - UI Name: Sliding window size\n",
      "        - Type: int\n",
      "        - Default Value: 4\n",
      "      - `sliding_window_min_quality`: \n",
      "        - UI Name: Sliding window minimum quality\n",
      "        - Type: int\n",
      "        - Default Value: 15\n",
      "```\u001b[00m\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mWhat is the name of the input object with UPA '211157/2/1'?\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mGet object name\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"upa\\\": \\\"211157/2/1\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "rhodo.art.q20.PE.reads\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "The name of the input object with UPA '211157/2/1' is 'rhodo.art.q20.PE.reads'.\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mTo populate the dictionary with the appropriate parameter values, I need to gather information about the input object name for the UPA \"211157/2/1\" and ensure that all non-optional parameters have values. I will start by retrieving the input object name using the Workspace Manager.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mAsk question to coworker\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"question\\\": \\\"What is the name of the input object with UPA '211157/2/1'?\\\", \\\"context\\\": \\\"I need to know the name of the input object with UPA '211157/2/1' to populate the 'output_reads_name' parameter for the KBase app 'kb_trimmomatic/run_trimmomatic'.\\\", \\\"coworker\\\": \\\"Workspace Manager\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "The name of the input object with UPA '211157/2/1' is 'rhodo.art.q20.PE.reads'.\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"narrative_id\": 211157,\n",
      "  \"app_id\": \"kb_trimmomatic/run_trimmomatic\",\n",
      "  \"app_params\": {\n",
      "    \"input_reads_ref\": \"211157/2/1\",\n",
      "    \"output_reads_name\": \"rhodo_art_q20_PE_reads_trimmed\",\n",
      "    \"translate_to_phred33\": 1,\n",
      "    \"crop_length\": null,\n",
      "    \"head_crop_length\": 0,\n",
      "    \"leading_min_quality\": 3,\n",
      "    \"trailing_min_quality\": 3,\n",
      "    \"min_length\": 36,\n",
      "    \"adapter_clip\": {\n",
      "      \"adapterFa\": \"TruSeq3-PE\",\n",
      "      \"seed_mismatches\": 2,\n",
      "      \"palindrome_clip_threshold\": 30,\n",
      "      \"simple_clip_threshold\": 10\n",
      "    },\n",
      "    \"sliding_window\": {\n",
      "      \"sliding_window_size\": 4,\n",
      "      \"sliding_window_min_quality\": 15\n",
      "    }\n",
      "  }\n",
      "}\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Using the app parameters and app id, use provided tools to start a new KBase app, which will return a job id.\n",
      "            Use that job id to create a new App Cell in the narrative such that narrative_id=211157. Return only the\n",
      "            job id.\n",
      "            \u001b[00m\n",
      "starting start_job tool\n",
      "narrative_id: 211157\n",
      "app_id: kb_trimmomatic/run_trimmomatic\n",
      "params: {'input_reads_ref': '211157/2/1', 'output_reads_name': 'rhodo_art_q20_PE_reads_trimmed', 'translate_to_phred33': 1, 'crop_length': None, 'head_crop_length': 0, 'leading_min_quality': 3, 'trailing_min_quality': 3, 'min_length': 36, 'adapter_clip': {'adapterFa': 'TruSeq3-PE', 'seed_mismatches': 2, 'palindrome_clip_threshold': 30, 'simple_clip_threshold': 10}, 'sliding_window': {'sliding_window_size': 4, 'sliding_window_min_quality': 15}}\n",
      "starting JobAgent._start_job\n",
      "narrative_id: 211157\n",
      "app_id: kb_trimmomatic/run_trimmomatic\n",
      "params: {'input_reads_ref': '211157/2/1', 'output_reads_name': 'rhodo_art_q20_PE_reads_trimmed', 'translate_to_phred33': 1, 'crop_length': None, 'head_crop_length': 0, 'leading_min_quality': 3, 'trailing_min_quality': 3, 'min_length': 36, 'adapter_clip': {'adapterFa': 'TruSeq3-PE', 'seed_mismatches': 2, 'palindrome_clip_threshold': 30, 'simple_clip_threshold': 10}, 'sliding_window': {'sliding_window_size': 4, 'sliding_window_min_quality': 15}}\n",
      "{'method': 'kb_trimmomatic.runTrimmomatic', 'service_ver': '5fb0fc6aa22ac62359f5bd8f0d407f51c99beea5', 'params': [{'input_ws': 'wjriehl:narrative_1742916273434', 'input_reads_ref': '211157/2/1', 'output_reads_name': 'rhodo_art_q20_PE_reads_trimmed', 'translate_to_phred33': 1, 'adapter_clip': {'adapterFa': 'TruSeq3-PE', 'seed_mismatches': 2, 'palindrome_clip_threshold': 30, 'simple_clip_threshold': 10}, 'sliding_window': {'sliding_window_size': 4, 'sliding_window_min_quality': 15}, 'leading_min_quality': 3, 'trailing_min_quality': 3, 'crop_length': None, 'head_crop_length': 0, 'min_length': 36}], 'app_id': 'kb_trimmomatic/run_trimmomatic', 'wsid': 211157, 'meta': {'cell_id': '431a2d36-b8a3-46e6-aa34-ab7b05504aac', 'run_id': '1fd7b455-8ce3-4d37-acb2-b3e3ad7dd3d3', 'tag': 'release'}, 'source_ws_objects': ['211157/2/1']}\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mstart-job\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"app_id\\\": \\\"kb_trimmomatic/run_trimmomatic\\\", \\\"params\\\": {\\\"input_reads_ref\\\": \\\"211157/2/1\\\", \\\"output_reads_name\\\": \\\"rhodo_art_q20_PE_reads_trimmed\\\", \\\"translate_to_phred33\\\": 1, \\\"crop_length\\\": null, \\\"head_crop_length\\\": 0, \\\"leading_min_quality\\\": 3, \\\"trailing_min_quality\\\": 3, \\\"min_length\\\": 36, \\\"adapter_clip\\\": {\\\"adapterFa\\\": \\\"TruSeq3-PE\\\", \\\"seed_mismatches\\\": 2, \\\"palindrome_clip_threshold\\\": 30, \\\"simple_clip_threshold\\\": 10}, \\\"sliding_window\\\": {\\\"sliding_window_size\\\": 4, \\\"sliding_window_min_quality\\\": 15}}}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "67e2f37b95ea352d72ffa06d\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "67e2f37b95ea352d72ffa06d\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Using the job id, create an app cell in narrative 211157. The add_app_cell tool is useful here.\n",
      "            \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92madd-app-cell\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"job_id\\\": \\\"67e2f37b95ea352d72ffa06d\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "success\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "success\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Using the job id, monitor the progress of the running job. The monitor_job tool is helpful here.\n",
      "            If completed, continue to the next task. If in an error state, summarize the error for the user and stop.\n",
      "            \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mThought: I need to monitor the job using the provided job ID to determine its final status.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mmonitor-job\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"job_id\\\": \\\"67e2f37b95ea352d72ffa06d\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "job_id='67e2f37b95ea352d72ffa06d' job_status='completed' created_objects=[CreatedObject(object_upa='211157/4/1', object_name='rhodo_art_q20_PE_reads_trimmed_paired')] job_error=None report_upa='211157/5/1'\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"job_id\": \"67e2f37b95ea352d72ffa06d\",\n",
      "  \"job_status\": \"completed\",\n",
      "  \"created_objects\": [\n",
      "    {\n",
      "      \"object_upa\": \"211157/4/1\",\n",
      "      \"object_name\": \"rhodo_art_q20_PE_reads_trimmed_paired\"\n",
      "    }\n",
      "  ],\n",
      "  \"job_error\": null,\n",
      "  \"report_upa\": \"211157/5/1\"\n",
      "}\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mGet the report UPA from the job results (a string with format number/number/number), an\n",
      "            use a tool to get the report object from the Workspace service. UPAs are unique permanent addresses used to identify data objects.\n",
      "            The Workspace Manager will be helpful here. Make sure the delegated agent uses a tool for interacting with KBase.\n",
      "            If there was an error in the job results, return a string saying so and summarizing the error.\n",
      "            If no report is available, either because there is no report UPA or the report object is unavailable, return a string saying so.\n",
      "            Otherwise, return the full report text. Your final answer MUST be the report text.\n",
      "            \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mget-report\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"upa\\\": \\\"211157/5/1\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "message: \n",
      "direct html: \n",
      "html report: rhodo_art_q20_PE_reads_trimmed HTML:\n",
      "<html>\n",
      "<body bgcolor=\"white\">\n",
      "<p><b><font color=\"#606060\">TRIMMOMATIC RESULTS FOR rhodo.art.q20.PE.reads (object 211157/2/1)</font></b><br>\n",
      "\n",
      "<table cellpadding=0 cellspacing=0 border=0>\n",
      "<tr><td></td><td>&nbsp;&nbsp;&nbsp;&nbsp;</td><td></td><td>&nbsp;&nbsp;</td></tr>\n",
      "<tr>\n",
      "    <td align=right><font color=\"#606060\">Input Read Pairs</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">193053</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">(100.0%)&nbsp;&nbsp;</font></td><td></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "</tr>\n",
      "<tr><td><font size=-2>&nbsp;</font></td></tr>\n",
      "<tr>\n",
      "    <td align=right><font color=\"#606060\">Both Surviving</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">193053</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">(100.0%)&nbsp;&nbsp;</font></td><td></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "    <td bgcolor=\"lightblue\"><font size=-2 color=\"lightblue\">.</font></td>\n",
      "</tr>\n",
      "<tr><td><font size=-2>&nbsp;</font></td></tr>\n",
      "<tr>\n",
      "    <td align=right><font color=\"#606060\">Forward Only Surviving</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">0</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">(0.0%)&nbsp;&nbsp;</font></td><td></td>\n",
      "</tr>\n",
      "<tr><td><font size=-2>&nbsp;</font></td></tr>\n",
      "<tr>\n",
      "    <td align=right><font color=\"#606060\">Reverse Only Surviving</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">0</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">(0.0%)&nbsp;&nbsp;</font></td><td></td>\n",
      "</tr>\n",
      "<tr><td><font size=-2>&nbsp;</font></td></tr>\n",
      "<tr>\n",
      "    <td align=right><font color=\"#606060\">Dropped</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">0</font></td><td></td>\n",
      "    <td align=right><font color=\"#606060\">(0.0%)&nbsp;&nbsp;</font></td><td></td>\n",
      "</tr>\n",
      "<tr><td><font size=-2>&nbsp;</font></td></tr>\n",
      "</table>\n",
      "<p>\n",
      "</body>\n",
      "</html>\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "TRIMMOMATIC RESULTS FOR rhodo.art.q20.PE.reads (object 211157/2/1)\n",
      "\n",
      "Input Read Pairs: 193053 (100.0%)\n",
      "Both Surviving: 193053 (100.0%)\n",
      "Forward Only Surviving: 0 (0.0%)\n",
      "Reverse Only Surviving: 0 (0.0%)\n",
      "Dropped: 0 (0.0%)\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mAnalyze the given report and derive some biological insight into the result.\n",
      "            If the report is not in JSON format, then interpret the document as-is.\n",
      "            If it is in JSON format, The report may contain content in 3 categories.\n",
      "            1. \"message\": this is a brief message describing the outcome of the report\n",
      "            2. \"direct html\": this is HTML-formatted information meant to be displayed to the user, and might be a brief summary of the full report.\n",
      "            3. \"html report\": this is one or more full HTML-formatted documents containing the report information.\n",
      "\n",
      "            Regardless of the format, it may be plain text, which can be interpreted as-is. It may also\n",
      "            also be formatted as HTML. If so, read the HTML document, including any base-64 encoded images for interpretation.\n",
      "\n",
      "            After interpretation, analyze the report, and summarize the findings with a biological interpretation. Write a brief summary, including\n",
      "            bullet points when appropriate, formatted in markdown. Your final answer MUST be the summary of the report.\n",
      "\n",
      "            If there was no report, the report is null, or an empty string, return a note saying that there is no report to analyze.\n",
      "            If the previous task ended with an error, or another note saying that there is no report to analyze,\n",
      "            just return a note saying so. Otherwise, return the summary of the report.\n",
      "            \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "### Trimmomatic Analysis Summary for rhodo.art.q20.PE.reads\n",
      "\n",
      "- **Input Read Pairs**: 193,053 (100.0%)\n",
      "- **Both Surviving**: 193,053 (100.0%)\n",
      "- **Forward Only Surviving**: 0 (0.0%)\n",
      "- **Reverse Only Surviving**: 0 (0.0%)\n",
      "- **Dropped**: 0 (0.0%)\n",
      "\n",
      "#### Biological Interpretation:\n",
      "- The Trimmomatic analysis indicates that all input read pairs were retained after processing, with no reads being dropped or surviving as single-end reads.\n",
      "- This suggests that the quality of the input data was high, with no need for trimming or filtering of low-quality reads.\n",
      "- The results imply that the dataset is well-suited for downstream analyses, such as assembly or variant calling, as the integrity of the paired-end reads is fully maintained.\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mSave the analysis by adding a markdown cell to the Narrative with id 211157. The markdown text must\n",
      "            be the analysis text. If not successful, say so and stop. If an output object was created, ensure that it has both an UPA and name.\n",
      "            In the end, return the results of the job completion task with both UPA and name for output object. The return result must be normalized to contain\n",
      "            both the UPA and output object name, if either are present. If the app has neither an output object name or UPA, both of these\n",
      "            fields may be None.\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92madd-markdown-cell\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"markdown_text\\\": \\\"### Trimmomatic Analysis Summary for rhodo.art.q20.PE.reads\\\\n\\\\n- **Input Read Pairs**: 193,053 (100.0%)\\\\n- **Both Surviving**: 193,053 (100.0%)\\\\n- **Forward Only Surviving**: 0 (0.0%)\\\\n- **Reverse Only Surviving**: 0 (0.0%)\\\\n- **Dropped**: 0 (0.0%)\\\\n\\\\n#### Biological Interpretation:\\\\n- The Trimmomatic analysis indicates that all input read pairs were retained after processing, with no reads being dropped or surviving as single-end reads.\\\\n- This suggests that the quality of the input data was high, with no need for trimming or filtering of low-quality reads.\\\\n- The results imply that the dataset is well-suited for downstream analyses, such as assembly or variant calling, as the integrity of the paired-end reads is fully maintained.\\\\n\\\\n----------\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "success\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"job_id\": \"67e2f37b95ea352d72ffa06d\",\n",
      "  \"job_status\": \"completed\",\n",
      "  \"created_objects\": [\n",
      "    {\n",
      "      \"object_upa\": \"211157/4/1\",\n",
      "      \"object_name\": \"rhodo_art_q20_PE_reads_trimmed_paired\"\n",
      "    }\n",
      "  ],\n",
      "  \"job_error\": null,\n",
      "  \"report_upa\": \"211157/5/1\"\n",
      "}\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mYou are a workflow runner, your role is to efficiently run KBase workflows.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mRun KBase app\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"app_id\\\": \\\"kb_trimmomatic/run_trimmomatic\\\", \\\"input_object_upa\\\": \\\"211157/2/1\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "job_id='67e2f37b95ea352d72ffa06d' job_status='completed' created_objects=[CreatedObject(object_upa='211157/4/1', object_name='rhodo_art_q20_PE_reads_trimmed_paired')] job_error=None report_upa='211157/5/1'\u001b[00m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Overriding of current TracerProvider is not allowed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            From the given KBase app id, kb_SPAdes/run_SPAdes, fetch the list of parameters needed to run it. Use the App and Job manager agent\n",
      "            for assistance. Using the data object with UPA \"211157/4/1\", populate a dictionary\n",
      "            with the parameters where the keys are parameter ids, and values are the proper parameter values, or their\n",
      "            default values if no value can be found or calculated.\n",
      "            Any input object parameter must be the input object UPA.\n",
      "            Be sure to make sure there is a non-null value for any parameter that is not optional.\n",
      "            Any parameter that has a true value for \"is_output_object\" must have a valid name for the new object. The new object name should be based on\n",
      "            the input object name, not its UPA. If the input object name is not available, the Workspace Manager can assist.\n",
      "            If the parameter type is 'dropdown', use the allowed 'name' option to determine what should be used, but only set the\n",
      "            'value' associated with that name, or the default value if any.\n",
      "            Only alphanumeric characters and underscores are allowed in new object names.\n",
      "            Return the dictionary of inputs, the app id, and the\n",
      "            narrative id 211157 for use in the next task. Do not add comments or other text. If the parameters are rejected, examine the reason why and\n",
      "            reform them. The dictionary of inputs and the app id must not be combined into a single dictionary.\n",
      "            \u001b[00m\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mCan you provide the list of parameters needed to run the KBase app with the id 'kb_SPAdes/run_SPAdes'?\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mget-app-parameters\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"app_id\\\": \\\"kb_SPAdes/run_SPAdes\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "{\"read_libraries\": {\"id\": \"read_libraries\", \"ui_name\": \"Read library\", \"short_hint\": \"The libraries of reads to be assembled.\", \"is_output_object\": false, \"type\": \"data_object\", \"allowed_data_type\": [\"KBaseFile.PairedEndLibrary\", \"KBaseAssembly.PairedEndLibrary\", \"KBaseFile.SingleEndLibrary\", \"KBaseAssembly.SingleEndLibrary\"], \"multiple\": true, \"optional\": false, \"default_value\": null}, \"dna_source\": {\"id\": \"dna_source\", \"ui_name\": \"DNA source\", \"short_hint\": \"The source of the sequenced DNA - either standard, single cell /\\nMDA, or plasmid.\", \"is_output_object\": false, \"type\": \"dropdown\", \"allowed\": [{\"name\": \"standard\", \"value\": \"standard\"}, {\"name\": \"single cell\", \"value\": \"single_cell\"}, {\"name\": \"plasmid\", \"value\": \"plasmid\"}], \"multiple\": false, \"optional\": false, \"default_value\": \"standard\"}, \"output_contigset_name\": {\"id\": \"output_contigset_name\", \"ui_name\": \"Output ContigSet\", \"short_hint\": \"The name under which the resulting ContigSet will be saved in the Narrative.\", \"is_output_object\": true, \"type\": \"data_object\", \"allowed_data_type\": [\"KBaseGenomeAnnotations.Assembly\"], \"multiple\": false, \"optional\": false, \"default_value\": \"SPAdes.Assembly\"}, \"min_contig_length\": {\"id\": \"min_contig_length\", \"ui_name\": \"Minimum Contig Length\", \"short_hint\": \"The shortest contig to accept in the resulting assembly object.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [200, Infinity], \"multiple\": false, \"optional\": false, \"default_value\": 500}, \"kmer_sizes\": {\"id\": \"kmer_sizes\", \"ui_name\": \"K-mer Sizes\", \"short_hint\": \"K-mer size values must be odd, less than 128, and listed in ascending order. In the absence of this input, values for K are automatically selected.\", \"is_output_object\": false, \"type\": \"int\", \"allowed_range\": [-Infinity, Infinity], \"multiple\": true, \"optional\": true, \"default_value\": null}, \"skip_error_correction\": {\"id\": \"skip_error_correction\", \"ui_name\": \"Assembly only (No Error Correction)\", \"short_hint\": \"Assembly only (No Error Correction).\", \"is_output_object\": false, \"type\": \"checkbox\", \"allowed\": [], \"multiple\": false, \"optional\": true, \"default_value\": \"0\"}}\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"read_libraries\": {\n",
      "    \"id\": \"read_libraries\",\n",
      "    \"ui_name\": \"Read library\",\n",
      "    \"short_hint\": \"The libraries of reads to be assembled.\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"data_object\",\n",
      "    \"allowed_data_type\": [\n",
      "      \"KBaseFile.PairedEndLibrary\",\n",
      "      \"KBaseAssembly.PairedEndLibrary\",\n",
      "      \"KBaseFile.SingleEndLibrary\",\n",
      "      \"KBaseAssembly.SingleEndLibrary\"\n",
      "    ],\n",
      "    \"multiple\": true,\n",
      "    \"optional\": false,\n",
      "    \"default_value\": null\n",
      "  },\n",
      "  \"dna_source\": {\n",
      "    \"id\": \"dna_source\",\n",
      "    \"ui_name\": \"DNA source\",\n",
      "    \"short_hint\": \"The source of the sequenced DNA - either standard, single cell /\\nMDA, or plasmid.\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"dropdown\",\n",
      "    \"allowed\": [\n",
      "      {\n",
      "        \"name\": \"standard\",\n",
      "        \"value\": \"standard\"\n",
      "      },\n",
      "      {\n",
      "        \"name\": \"single cell\",\n",
      "        \"value\": \"single_cell\"\n",
      "      },\n",
      "      {\n",
      "        \"name\": \"plasmid\",\n",
      "        \"value\": \"plasmid\"\n",
      "      }\n",
      "    ],\n",
      "    \"multiple\": false,\n",
      "    \"optional\": false,\n",
      "    \"default_value\": \"standard\"\n",
      "  },\n",
      "  \"output_contigset_name\": {\n",
      "    \"id\": \"output_contigset_name\",\n",
      "    \"ui_name\": \"Output ContigSet\",\n",
      "    \"short_hint\": \"The name under which the resulting ContigSet will be saved in the Narrative.\",\n",
      "    \"is_output_object\": true,\n",
      "    \"type\": \"data_object\",\n",
      "    \"allowed_data_type\": [\n",
      "      \"KBaseGenomeAnnotations.Assembly\"\n",
      "    ],\n",
      "    \"multiple\": false,\n",
      "    \"optional\": false,\n",
      "    \"default_value\": \"SPAdes.Assembly\"\n",
      "  },\n",
      "  \"min_contig_length\": {\n",
      "    \"id\": \"min_contig_length\",\n",
      "    \"ui_name\": \"Minimum Contig Length\",\n",
      "    \"short_hint\": \"The shortest contig to accept in the resulting assembly object.\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"int\",\n",
      "    \"allowed_range\": [\n",
      "      200,\n",
      "      Infinity\n",
      "    ],\n",
      "    \"multiple\": false,\n",
      "    \"optional\": false,\n",
      "    \"default_value\": 500\n",
      "  },\n",
      "  \"kmer_sizes\": {\n",
      "    \"id\": \"kmer_sizes\",\n",
      "    \"ui_name\": \"K-mer Sizes\",\n",
      "    \"short_hint\": \"K-mer size values must be odd, less than 128, and listed in ascending order. In the absence of this input, values for K are automatically selected.\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"int\",\n",
      "    \"allowed_range\": [\n",
      "      -Infinity,\n",
      "      Infinity\n",
      "    ],\n",
      "    \"multiple\": true,\n",
      "    \"optional\": true,\n",
      "    \"default_value\": null\n",
      "  },\n",
      "  \"skip_error_correction\": {\n",
      "    \"id\": \"skip_error_correction\",\n",
      "    \"ui_name\": \"Assembly only (No Error Correction)\",\n",
      "    \"short_hint\": \"Assembly only (No Error Correction).\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"checkbox\",\n",
      "    \"allowed\": [],\n",
      "    \"multiple\": false,\n",
      "    \"optional\": true,\n",
      "    \"default_value\": \"0\"\n",
      "  }\n",
      "}\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mTo complete the task, I need to first fetch the list of parameters needed to run the KBase app with the id \"kb_SPAdes/run_SPAdes\" using the Job and App Manager. Once I have the list of parameters, I can then populate a dictionary with the appropriate values, ensuring that all non-optional parameters have non-null values and that any output object has a valid name.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mAsk question to coworker\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"question\\\": \\\"Can you provide the list of parameters needed to run the KBase app with the id 'kb_SPAdes/run_SPAdes'?\\\", \\\"context\\\": \\\"I need to populate a dictionary with these parameters for running the app, ensuring all non-optional parameters have non-null values and any output object has a valid name.\\\", \\\"coworker\\\": \\\"Job and App Manager\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"read_libraries\": {\n",
      "    \"id\": \"read_libraries\",\n",
      "    \"ui_name\": \"Read library\",\n",
      "    \"short_hint\": \"The libraries of reads to be assembled.\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"data_object\",\n",
      "    \"allowed_data_type\": [\n",
      "      \"KBaseFile.PairedEndLibrary\",\n",
      "      \"KBaseAssembly.PairedEndLibrary\",\n",
      "      \"KBaseFile.SingleEndLibrary\",\n",
      "      \"KBaseAssembly.SingleEndLibrary\"\n",
      "    ],\n",
      "    \"multiple\": true,\n",
      "    \"optional\": false,\n",
      "    \"default_value\": null\n",
      "  },\n",
      "  \"dna_source\": {\n",
      "    \"id\": \"dna_source\",\n",
      "    \"ui_name\": \"DNA source\",\n",
      "    \"short_hint\": \"The source of the sequenced DNA - either standard, single cell /\\nMDA, or plasmid.\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"dropdown\",\n",
      "    \"allowed\": [\n",
      "      {\n",
      "        \"name\": \"standard\",\n",
      "        \"value\": \"standard\"\n",
      "      },\n",
      "      {\n",
      "        \"name\": \"single cell\",\n",
      "        \"value\": \"single_cell\"\n",
      "      },\n",
      "      {\n",
      "        \"name\": \"plasmid\",\n",
      "        \"value\": \"plasmid\"\n",
      "      }\n",
      "    ],\n",
      "    \"multiple\": false,\n",
      "    \"optional\": false,\n",
      "    \"default_value\": \"standard\"\n",
      "  },\n",
      "  \"output_contigset_name\": {\n",
      "    \"id\": \"output_contigset_name\",\n",
      "    \"ui_name\": \"Output ContigSet\",\n",
      "    \"short_hint\": \"The name under which the resulting ContigSet will be saved in the Narrative.\",\n",
      "    \"is_output_object\": true,\n",
      "    \"type\": \"data_object\",\n",
      "    \"allowed_data_type\": [\n",
      "      \"KBaseGenomeAnnotations.Assembly\"\n",
      "    ],\n",
      "    \"multiple\": false,\n",
      "    \"optional\": false,\n",
      "    \"default_value\": \"SPAdes.Assembly\"\n",
      "  },\n",
      "  \"min_contig_length\": {\n",
      "    \"id\": \"min_contig_length\",\n",
      "    \"ui_name\": \"Minimum Contig Length\",\n",
      "    \"short_hint\": \"The shortest contig to accept in the resulting assembly object.\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"int\",\n",
      "    \"allowed_range\": [\n",
      "      200,\n",
      "      Infinity\n",
      "    ],\n",
      "    \"multiple\": false,\n",
      "    \"optional\": false,\n",
      "    \"default_value\": 500\n",
      "  },\n",
      "  \"kmer_sizes\": {\n",
      "    \"id\": \"kmer_sizes\",\n",
      "    \"ui_name\": \"K-mer Sizes\",\n",
      "    \"short_hint\": \"K-mer size values must be odd, less than 128, and listed in ascending order. In the absence of this input, values for K are automatically selected.\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"int\",\n",
      "    \"allowed_range\": [\n",
      "      -Infinity,\n",
      "      Infinity\n",
      "    ],\n",
      "    \"multiple\": true,\n",
      "    \"optional\": true,\n",
      "    \"default_value\": null\n",
      "  },\n",
      "  \"skip_error_correction\": {\n",
      "    \"id\": \"skip_error_correction\",\n",
      "    \"ui_name\": \"Assembly only (No Error Correction)\",\n",
      "    \"short_hint\": \"Assembly only (No Error Correction).\",\n",
      "    \"is_output_object\": false,\n",
      "    \"type\": \"checkbox\",\n",
      "    \"allowed\": [],\n",
      "    \"multiple\": false,\n",
      "    \"optional\": true,\n",
      "    \"default_value\": \"0\"\n",
      "  }\n",
      "}\n",
      "```\u001b[00m\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mWhat is the name of the input object with UPA '211157/4/1'?\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mI need to retrieve the name of the object with the given UPA '211157/4/1' to provide the required information.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mGet object name\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"upa\\\": \\\"211157/4/1\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "rhodo_art_q20_PE_reads_trimmed_paired\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mWorkspace Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "The name of the input object with UPA '211157/4/1' is 'rhodo_art_q20_PE_reads_trimmed_paired'.\u001b[00m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mTo proceed, I need to gather the input object name for the UPA \"211157/4/1\" to ensure the output object name is valid. I will ask the Workspace Manager for assistance with this.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mAsk question to coworker\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"question\\\": \\\"What is the name of the input object with UPA '211157/4/1'?\\\", \\\"context\\\": \\\"I need the input object name to create a valid output object name for the KBase app 'kb_SPAdes/run_SPAdes'.\\\", \\\"coworker\\\": \\\"Workspace Manager\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "The name of the input object with UPA '211157/4/1' is 'rhodo_art_q20_PE_reads_trimmed_paired'.\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mProject coordinator\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"narrative_id\": 211157,\n",
      "  \"app_id\": \"kb_SPAdes/run_SPAdes\",\n",
      "  \"app_params\": {\n",
      "    \"read_libraries\": [\"211157/4/1\"],\n",
      "    \"dna_source\": \"standard\",\n",
      "    \"output_contigset_name\": \"rhodo_art_q20_PE_reads_trimmed_paired_Assembly\",\n",
      "    \"min_contig_length\": 500,\n",
      "    \"kmer_sizes\": null,\n",
      "    \"skip_error_correction\": \"0\"\n",
      "  }\n",
      "}\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Using the app parameters and app id, use provided tools to start a new KBase app, which will return a job id.\n",
      "            Use that job id to create a new App Cell in the narrative such that narrative_id=211157. Return only the\n",
      "            job id.\n",
      "            \u001b[00m\n",
      "starting start_job tool\n",
      "narrative_id: 211157\n",
      "app_id: kb_SPAdes/run_SPAdes\n",
      "params: {'read_libraries': ['211157/4/1'], 'dna_source': 'standard', 'output_contigset_name': 'rhodo_art_q20_PE_reads_trimmed_paired_Assembly', 'min_contig_length': 500, 'kmer_sizes': None, 'skip_error_correction': '0'}\n",
      "starting JobAgent._start_job\n",
      "narrative_id: 211157\n",
      "app_id: kb_SPAdes/run_SPAdes\n",
      "params: {'read_libraries': ['211157/4/1'], 'dna_source': 'standard', 'output_contigset_name': 'rhodo_art_q20_PE_reads_trimmed_paired_Assembly', 'min_contig_length': 500, 'kmer_sizes': None, 'skip_error_correction': '0'}\n",
      "{'method': 'kb_SPAdes.run_SPAdes', 'service_ver': 'ee3f83f9d451a7d0ad86709e55e8653bee497e30', 'params': [{'workspace_name': 'wjriehl:narrative_1742916273434', 'read_libraries': ['211157/4/1'], 'output_contigset_name': 'rhodo_art_q20_PE_reads_trimmed_paired_Assembly', 'dna_source': 'standard', 'min_contig_length': 500, 'kmer_sizes': None, 'skip_error_correction': '0'}], 'app_id': 'kb_SPAdes/run_SPAdes', 'wsid': 211157, 'meta': {'cell_id': 'd62dbb81-770d-4b49-b092-518758e7059d', 'run_id': '8531f7c7-8348-43bc-b5f3-18c1e6228138', 'tag': 'release'}, 'source_ws_objects': ['211157/4/1']}\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Thought:\u001b[00m \u001b[92mI have the necessary information to start a new KBase app. I will use the `start-job` tool with the provided narrative_id, app_id, and app_params to start the job.\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92mstart-job\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"app_id\\\": \\\"kb_SPAdes/run_SPAdes\\\", \\\"params\\\": {\\\"read_libraries\\\": [\\\"211157/4/1\\\"], \\\"dna_source\\\": \\\"standard\\\", \\\"output_contigset_name\\\": \\\"rhodo_art_q20_PE_reads_trimmed_paired_Assembly\\\", \\\"min_contig_length\\\": 500, \\\"kmer_sizes\\\": null, \\\"skip_error_correction\\\": \\\"0\\\"}}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "67e2f4361b77ed423827fd03\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "67e2f4361b77ed423827fd03\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Using the job id, create an app cell in narrative 211157. The add_app_cell tool is useful here.\n",
      "            \u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Using tool:\u001b[00m \u001b[92madd-app-cell\u001b[00m\n",
      "\u001b[95m## Tool Input:\u001b[00m \u001b[92m\n",
      "\"{\\\"narrative_id\\\": 211157, \\\"job_id\\\": \\\"67e2f4361b77ed423827fd03\\\"}\"\u001b[00m\n",
      "\u001b[95m## Tool Output:\u001b[00m \u001b[92m\n",
      "success\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNarrative Manager\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "success\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mJob and App Manager\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92m\n",
      "            Using the job id, monitor the progress of the running job. The monitor_job tool is helpful here.\n",
      "            If completed, continue to the next task. If in an error state, summarize the error for the user and stop.\n",
      "            \u001b[00m\n"
     ]
    }
   ],
   "source": [
    "crew = Crew(\n",
    "    agents = [wf_runner.agent],\n",
    "    tasks=[run_apps_task],\n",
    "    verbose=True,\n",
    ")\n",
    "crew.kickoff()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "852137e9-2c9e-4cd2-b541-0bbdfc52d25d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
